{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Simulating ancestry with msprime\n",
    "\n",
    "10.30am\n",
    "\n",
    "This is the first of two sessions about `msprime`, a backwards-time tree sequence simulator. In this first worksheet, we will learn how to design simulations using customisable models of genomic history. In the next session, we will add mutations to these simulated histories to obtain sequence data.\n",
    "\n",
    "#### Some relevant papers:\n",
    " -  [Efficient coalescent simulation and genealogical analysis for large sample sizes](https://doi.org/10.1371/journal.pcbi.1004842)\n",
    " - [Efficient ancestry and mutation simulation with msprime 1.0](https://doi.org/10.1093/genetics/iyab229)\n",
    " - [tskit.dev documentation](https://tskit.dev/)\n",
    " \n",
    "Simulations are important in population genetics for many reasons:\n",
    "\n",
    "**Exploration:**\n",
    "Simulations allow us to explore the influence of various historical scenarios on observed patterns of genetic variation and inheritance.\n",
    "\n",
    "**Benchmarking and evaluating methodologies:**\n",
    "To assess the accuracy of inferential methods, we need test datasets for which the true values of important parameters are known.\n",
    "\n",
    "**Model training:**\n",
    "Some methods for ancestry inference are trained on simulated data (eg. Approximate Bayesian Computation).\n",
    "This is especially important in studies of complex demographies, where there are many potential parameters and models, making it impractical to specify likelihood functions.\n",
    "\n",
    "### A brief history of msprime\n",
    "\n",
    "The first release of `msprime` was an emulation of the popular `ms` coalescent simulator with added support for tree sequences.\n",
    "However, it has since become an expansive and flexible backwards-in-time simulator for various different models of genetic ancestry and mutation, and even for simplified models of selection.\n",
    "\n",
    "#### Forwards and backwards simulation\n",
    "\n",
    "The main characteristic of `msprime` is that is simulates *tree sequences* in *backwards-time*.\n",
    "\n",
    "<img src=\"pics/msprime-1.png\" width=\"200\" height=\"200\">\n",
    "<img src=\"pics/msprime-2.png\" width=\"200\" height=\"200\">\n",
    "<img src=\"pics/msprime-3.png\" width=\"200\" height=\"200\">\n",
    "<img src=\"pics/msprime-4.png\" width=\"200\" height=\"200\">\n",
    "<img src=\"pics/msprime-5.png\" width=\"200\" height=\"200\">\n",
    "\n",
    "The alternative is to use a *forwards-time* simulator like `SLiM`:\n",
    "\n",
    "<img src=\"pics/slim-1.png\" width=\"200\" height=\"200\">\n",
    "<img src=\"pics/slim-2.png\" width=\"200\" height=\"200\">\n",
    "<img src=\"pics/slim-3.png\" width=\"200\" height=\"200\">\n",
    "<img src=\"pics/slim-4.png\" width=\"200\" height=\"200\">\n",
    "<img src=\"pics/slim-5.png\" width=\"200\" height=\"200\">\n",
    "<img src=\"pics/slim-6.png\" width=\"200\" height=\"200\">\n",
    "<img src=\"pics/slim-7.png\" width=\"200\" height=\"200\">\n",
    "<img src=\"pics/slim-8.png\" width=\"200\" height=\"200\">\n",
    "<img src=\"pics/slim-9.png\" width=\"200\" height=\"200\">\n",
    "<img src=\"pics/slim-10.png\" width=\"200\" height=\"200\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In general, forwards-time simulation with `SLiM` is *detailed* and *more realistic*,\n",
    "while backwards-time simulation with `msprime` is *fast* and *efficient*.\n",
    "`msprime` is designed for *neutral* simulations, while `SLiM` is better suited to complex simulations, including those involving selection.\n",
    "\n",
    "However, `pyslim` allows you to combine these via a process called 'recapitation'.\n",
    "See (link) if this is of interest.\n",
    "\n",
    "<img src=\"pics/recapitation-1.png\" width=\"400\" height=\"400\">\n",
    "<img src=\"pics/recapitation-2.png\" width=\"400\" height=\"400\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import msprime\n",
    "import random\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from IPython.display import SVG"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Basic syntax\n",
    "\n",
    "(10.40am)\n",
    "\n",
    "To simulate a tree sequence in `msprime`, we use the `sim_ancestry()` method:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n6 p0 root\" transform=\"translate(98.75 26.8)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(26.25 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H -26.25\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(-26.25 25.2505)\"><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(22.5 69.3495)\"><path class=\"edge\" d=\"M 0 0 V -69.3495 H -22.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a5 c2 node n4 p0\" transform=\"translate(-22.5 63.0417)\"><g class=\"a4 i0 leaf node n0 p0 sample\" transform=\"translate(-15 6.3078)\"><path class=\"edge\" d=\"M 0 0 V -6.3078 H 15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(15 6.3078)\"><path class=\"edge\" d=\"M 0 0 V -6.3078 H -15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><path class=\"edge\" d=\"M 0 0 V -63.0417 H 22.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -25.2505 H 26.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(samples=2, random_seed=1)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "              <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "              </style>\n",
       "              <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"padding:0;line-height:21px;\">\n",
       "                          <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                          <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                        </th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                      <tr><td>Trees</td><td>1</td></tr>\n",
       "                      <tr><td>Sequence Length</td><td>1.0</td></tr>\n",
       "                      <tr><td>Time Units</td><td>generations</td></tr>\n",
       "                      <tr><td>Sample Nodes</td><td>4</td></tr>\n",
       "                      <tr><td>Total Size</td><td>1.8 KiB</td></tr>\n",
       "                      <tr>\n",
       "                        <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"line-height:21px;\">Table</th>\n",
       "                        <th>Rows</th>\n",
       "                        <th>Size</th>\n",
       "                        <th>Has Metadata</th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                    \n",
       "                  <tr>\n",
       "                    <td>Edges</td>\n",
       "                      <td>6</td>\n",
       "                      <td>200 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Individuals</td>\n",
       "                      <td>2</td>\n",
       "                      <td>80 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Migrations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>8 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Mutations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>16 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Nodes</td>\n",
       "                      <td>7</td>\n",
       "                      <td>204 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Populations</td>\n",
       "                      <td>1</td>\n",
       "                      <td>224 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        âœ…\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Provenances</td>\n",
       "                      <td>1</td>\n",
       "                      <td>980 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Sites</td>\n",
       "                      <td>0</td>\n",
       "                      <td>16 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "              </div>\n",
       "            </div>\n",
       "            "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x108fd8b38>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Specifying contig information\n",
    "Although we have specified 2 samples, our tree sequence contains 4 sample nodes.\n",
    "This is because the `samples` argument specifies the number of *individuals* in the sample,\n",
    "and by default, `sim_ancestry()` assumes diploid organisms.\n",
    "To change this, use the `ploidy` argument:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n10 p0 root\" transform=\"translate(85 26.8)\"><g class=\"a10 c2 node n7 p0\" transform=\"translate(-30 50.894)\"><g class=\"a7 i1 leaf node n5 p0 sample\" transform=\"translate(15 43.706)\"><path class=\"edge\" d=\"M 0 0 V -43.706 H -15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><g class=\"a7 c2 node n6 p0\" transform=\"translate(-15 38.5615)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-10 5.14449)\"><path class=\"edge\" d=\"M 0 0 V -5.14449 H 10\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i0 leaf node n2 p0 sample\" transform=\"translate(10 5.14449)\"><path class=\"edge\" d=\"M 0 0 V -5.14449 H -10\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -38.5615 H 15\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -50.894 H 30\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><g class=\"a10 c2 node n9 p0\" transform=\"translate(30 29.4758)\"><g class=\"a9 i1 leaf node n3 p0 sample\" transform=\"translate(15 65.1242)\"><path class=\"edge\" d=\"M 0 0 V -65.1242 H -15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(-15 12.8374)\"><g class=\"a8 i0 leaf node n1 p0 sample\" transform=\"translate(-10 52.2867)\"><path class=\"edge\" d=\"M 0 0 V -52.2867 H 10\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a8 i1 leaf node n4 p0 sample\" transform=\"translate(10 52.2867)\"><path class=\"edge\" d=\"M 0 0 V -52.2867 H -10\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -12.8374 H 15\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -29.4758 H -30\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">9</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">10</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(samples=2, random_seed=1, ploidy=3)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It's easiest to start thinking about genome lengths in units of nucleotides. By default, we are simulating a sequence length that spans just one of these units.\n",
    "We can specify a larger region using the `sequence_length` argument:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">10</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n6 p0 root\" transform=\"translate(98.75 26.8)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(26.25 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H -26.25\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(-26.25 25.2505)\"><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(22.5 69.3495)\"><path class=\"edge\" d=\"M 0 0 V -69.3495 H -22.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a5 c2 node n4 p0\" transform=\"translate(-22.5 63.0417)\"><g class=\"a4 i0 leaf node n0 p0 sample\" transform=\"translate(-15 6.3078)\"><path class=\"edge\" d=\"M 0 0 V -6.3078 H 15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(15 6.3078)\"><path class=\"edge\" d=\"M 0 0 V -6.3078 H -15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><path class=\"edge\" d=\"M 0 0 V -63.0417 H 22.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -25.2505 H 26.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(samples=2, random_seed=1, sequence_length=10)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Also, note that our 'tree sequence' consists of just a single tree. This is because we have not yet specified a `recombination_rate`, and the default is 0. \n",
    "This is the probability of a recombination event per genomic unit (base), per generation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"800\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l190,0 l0,138.2 l-114,25 l0,5 l-76,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M210,0 l190,0 l0,138.2 l-228,25 l0,5 l-76,0 l0,-5 l114,-25 l0,-138.2z\"/><path d=\"M400,0 l190,0 l0,138.2 l114,25 l0,5 l-532,0 l0,-5 l228,-25 l0,-138.2z\"/><path d=\"M590,0 l190,0 l0,138.2 l0,25 l0,5 l-76,0 l0,-5 l-114,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(400 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"780\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(96 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g><g class=\"tick\" transform=\"translate(172 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">2</text></g></g><g class=\"tick\" transform=\"translate(704 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">9</text></g></g><g class=\"tick\" transform=\"translate(780 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">10</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n8 p0 root\" transform=\"translate(109.062 30.0081)\"><g class=\"a8 i1 leaf node n3 p0 sample\" transform=\"translate(42.1875 91.3919)\"><path class=\"edge\" d=\"M 0 0 V -91.3919 H -42.1875\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a8 c2 node n5 p0\" transform=\"translate(-42.1875 46.5549)\"><g class=\"a5 i0 leaf node n0 p0 sample\" transform=\"translate(-28.125 44.837)\"><path class=\"edge\" d=\"M 0 0 V -44.837 H 28.125\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a5 c2 node n4 p0\" transform=\"translate(28.125 44.1532)\"><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 0.683774)\"><path class=\"edge\" d=\"M 0 0 V -0.683774 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i1 leaf node n2 p0 sample\" transform=\"translate(18.75 0.683774)\"><path class=\"edge\" d=\"M 0 0 V -0.683774 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -44.1532 H -28.125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -46.5549 H 42.1875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">8</text></g></g></g><g class=\"tree t1\" transform=\"translate(210 0)\"><g class=\"plotbox\"><g class=\"c2 node n7 p0 root\" transform=\"translate(109.062 36.5039)\"><g class=\"a7 i1 leaf node n3 p0 sample\" transform=\"translate(42.1875 84.8961)\"><path class=\"edge\" d=\"M 0 0 V -84.8961 H -42.1875\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a7 c2 node n5 p0\" transform=\"translate(-42.1875 40.0591)\"><g class=\"a5 i0 leaf node n0 p0 sample\" transform=\"translate(-28.125 44.837)\"><path class=\"edge\" d=\"M 0 0 V -44.837 H 28.125\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a5 c2 node n4 p0\" transform=\"translate(28.125 44.1532)\"><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 0.683774)\"><path class=\"edge\" d=\"M 0 0 V -0.683774 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i1 leaf node n2 p0 sample\" transform=\"translate(18.75 0.683774)\"><path class=\"edge\" d=\"M 0 0 V -0.683774 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -44.1532 H -28.125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -40.0591 H 42.1875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">7</text></g></g></g><g class=\"tree t2\" transform=\"translate(400 0)\"><g class=\"plotbox\"><g class=\"c2 node n7 p0 root\" transform=\"translate(80.9375 36.5039)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-42.1875 84.8961)\"><path class=\"edge\" d=\"M 0 0 V -84.8961 H 42.1875\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 c2 node n6 p0\" transform=\"translate(42.1875 32.1058)\"><g class=\"a6 i1 leaf node n3 p0 sample\" transform=\"translate(28.125 52.7904)\"><path class=\"edge\" d=\"M 0 0 V -52.7904 H -28.125\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a6 c2 node n4 p0\" transform=\"translate(-28.125 52.1066)\"><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 0.683774)\"><path class=\"edge\" d=\"M 0 0 V -0.683774 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i1 leaf node n2 p0 sample\" transform=\"translate(18.75 0.683774)\"><path class=\"edge\" d=\"M 0 0 V -0.683774 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -52.1066 H 28.125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -32.1058 H -42.1875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">7</text></g></g></g><g class=\"tree t3\" transform=\"translate(590 0)\"><g class=\"plotbox\"><g class=\"c2 node n9 p0 root\" transform=\"translate(95 26.8)\"><g class=\"a9 c2 node n4 p0\" transform=\"translate(37.5 93.9162)\"><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 0.683774)\"><path class=\"edge\" d=\"M 0 0 V -0.683774 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i1 leaf node n2 p0 sample\" transform=\"translate(18.75 0.683774)\"><path class=\"edge\" d=\"M 0 0 V -0.683774 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -93.9162 H -37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4</text></g><g class=\"a9 c2 node n7 p0\" transform=\"translate(-37.5 9.70388)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-18.75 84.8961)\"><path class=\"edge\" d=\"M 0 0 V -84.8961 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 i1 leaf node n3 p0 sample\" transform=\"translate(18.75 84.8961)\"><path class=\"edge\" d=\"M 0 0 V -84.8961 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -9.70388 H 37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">9</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "                samples=2,\n",
    "                random_seed=82,\n",
    "                sequence_length=10,\n",
    "                recombination_rate=0.1)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We may also wish to specify recombination rates that vary across the genome. \n",
    "We do this by creating a `RateMap` object,\n",
    "which lists recombination rates between defined positions in the sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"800\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l190,0 l0,138.2 l228,25 l0,5 l-418,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M210,0 l190,0 l0,138.2 l76,25 l0,5 l-38,0 l0,-5 l-228,-25 l0,-138.2z\"/><path d=\"M400,0 l190,0 l0,138.2 l0,25 l0,5 l-114,0 l0,-5 l-76,-25 l0,-138.2z\"/><path d=\"M590,0 l190,0 l0,138.2 l0,25 l0,5 l-190,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(400 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"780\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(438 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">11</text></g></g><g class=\"tick\" transform=\"translate(476 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">12</text></g></g><g class=\"tick\" transform=\"translate(590 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">15</text></g></g><g class=\"tick\" transform=\"translate(780 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">20</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p0 root\" transform=\"translate(74.6875 26.8)\"><g class=\"a14 c2 node n6 p0\" transform=\"translate(-29.6875 93.0814)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.5 1.51859)\"><path class=\"edge\" d=\"M 0 0 V -1.51859 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.5 1.51859)\"><path class=\"edge\" d=\"M 0 0 V -1.51859 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -93.0814 H 29.6875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><g class=\"a14 c2 node n9 p0\" transform=\"translate(29.6875 68.193)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(-21.875 26.407)\"><path class=\"edge\" d=\"M 0 0 V -26.407 H 21.875\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(21.875 9.93916)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-18.75 16.4678)\"><path class=\"edge\" d=\"M 0 0 V -16.4678 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 c2 node n7 p0\" transform=\"translate(18.75 2.75709)\"><g class=\"a7 i1 leaf node n3 p0 sample\" transform=\"translate(-12.5 13.7107)\"><path class=\"edge\" d=\"M 0 0 V -13.7107 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a7 i2 leaf node n4 p0 sample\" transform=\"translate(12.5 13.7107)\"><path class=\"edge\" d=\"M 0 0 V -13.7107 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -2.75709 H -18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -9.93916 H -21.875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -68.193 H -29.6875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">9</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g><g class=\"tree t1\" transform=\"translate(210 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p0 root\" transform=\"translate(87.1875 26.8)\"><g class=\"a14 c2 node n6 p0\" transform=\"translate(-42.1875 93.0814)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.5 1.51859)\"><path class=\"edge\" d=\"M 0 0 V -1.51859 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.5 1.51859)\"><path class=\"edge\" d=\"M 0 0 V -1.51859 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -93.0814 H 42.1875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><g class=\"a14 c2 node n13 p0\" transform=\"translate(42.1875 13.747)\"><g class=\"a13 i1 leaf node n3 p0 sample\" transform=\"translate(28.125 80.853)\"><path class=\"edge\" d=\"M 0 0 V -80.853 H -28.125\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a13 c2 node n9 p0\" transform=\"translate(-28.125 54.446)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 26.407)\"><path class=\"edge\" d=\"M 0 0 V -26.407 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(18.75 9.93916)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.5 16.4678)\"><path class=\"edge\" d=\"M 0 0 V -16.4678 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.5 16.4678)\"><path class=\"edge\" d=\"M 0 0 V -16.4678 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -9.93916 H -18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -54.446 H 28.125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -13.747 H -42.1875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">13</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g><g class=\"tree t2\" transform=\"translate(400 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p0 root\" transform=\"translate(101.25 26.8)\"><g class=\"a14 c2 node n13 p0\" transform=\"translate(37.5 13.747)\"><g class=\"a13 i1 leaf node n3 p0 sample\" transform=\"translate(18.75 80.853)\"><path class=\"edge\" d=\"M 0 0 V -80.853 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a13 c2 node n8 p0\" transform=\"translate(-18.75 64.3851)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.5 16.4678)\"><path class=\"edge\" d=\"M 0 0 V -16.4678 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.5 16.4678)\"><path class=\"edge\" d=\"M 0 0 V -16.4678 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -64.3851 H 18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -13.747 H -37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">13</text></g><g class=\"a14 c2 node n10 p0\" transform=\"translate(-37.5 60.1721)\"><g class=\"a10 i0 leaf node n1 p0 sample\" transform=\"translate(18.75 34.4279)\"><path class=\"edge\" d=\"M 0 0 V -34.4279 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a10 c2 node n6 p0\" transform=\"translate(-18.75 32.9093)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.5 1.51859)\"><path class=\"edge\" d=\"M 0 0 V -1.51859 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.5 1.51859)\"><path class=\"edge\" d=\"M 0 0 V -1.51859 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -32.9093 H 18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -60.1721 H 37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">10</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g><g class=\"tree t3\" transform=\"translate(590 0)\"><g class=\"plotbox\"><g class=\"c2 node n13 p0 root\" transform=\"translate(95 40.547)\"><g class=\"a13 c2 node n11 p0\" transform=\"translate(31.25 16.6876)\"><g class=\"a11 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 64.1653)\"><path class=\"edge\" d=\"M 0 0 V -64.1653 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a11 c2 node n8 p0\" transform=\"translate(18.75 47.6975)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.5 16.4678)\"><path class=\"edge\" d=\"M 0 0 V -16.4678 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.5 16.4678)\"><path class=\"edge\" d=\"M 0 0 V -16.4678 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -47.6975 H -18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -16.6876 H -31.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">11</text></g><g class=\"a13 c2 node n12 p0\" transform=\"translate(-31.25 13.0746)\"><g class=\"a12 i1 leaf node n3 p0 sample\" transform=\"translate(18.75 67.7784)\"><path class=\"edge\" d=\"M 0 0 V -67.7784 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a12 c2 node n6 p0\" transform=\"translate(-18.75 66.2598)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.5 1.51859)\"><path class=\"edge\" d=\"M 0 0 V -1.51859 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.5 1.51859)\"><path class=\"edge\" d=\"M 0 0 V -1.51859 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -66.2598 H 18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -13.0746 H 31.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">12</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">13</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rate_map = msprime.RateMap(position=[0, 10, 20], rate=[0.01, 0.1])\n",
    "ts = msprime.sim_ancestry(3, recombination_rate=rate_map, random_seed=2)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that it's also possible to simulate gene conversion in `msprime`.\n",
    "If this is of interest, have a look at the API documentation for the `gene_conversion_rate` and `gene_conversion_tract_length` arguments,\n",
    "and [this](https://tskit.dev/msprime/docs/stable/ancestry.html?highlight=gene%20conversion#gene-conversion) short tutorial.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By default, the recombination events will be assigned to integer locations along the sequence. However there may be situations where you want to model the genome using continuous coordinates. In this case, use the `discrete_genome=False` argument:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"800\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l190,0 l0,138.2 l-43.4894,25 l0,5 l-146.511,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M210,0 l190,0 l0,138.2 l-72.5098,25 l0,5 l-160.98,0 l0,-5 l43.4894,-25 l0,-138.2z\"/><path d=\"M400,0 l190,0 l0,138.2 l94.7864,25 l0,5 l-357.296,0 l0,-5 l72.5098,-25 l0,-138.2z\"/><path d=\"M590,0 l190,0 l0,138.2 l0,25 l0,5 l-95.2136,0 l0,-5 l-94.7864,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(400 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"780\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0.00</text></g></g><g class=\"tick\" transform=\"translate(166.511 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">19.28</text></g></g><g class=\"tick\" transform=\"translate(327.49 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">40.46</text></g></g><g class=\"tick\" transform=\"translate(684.786 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">87.47</text></g></g><g class=\"tick\" transform=\"translate(780 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">100.00</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n7 p0 root\" transform=\"translate(80.9375 75.3217)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-42.1875 46.0783)\"><path class=\"edge\" d=\"M 0 0 V -46.0783 H 42.1875\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 c2 node n5 p0\" transform=\"translate(42.1875 41.0358)\"><g class=\"a5 i1 leaf node n2 p0 sample\" transform=\"translate(28.125 5.04243)\"><path class=\"edge\" d=\"M 0 0 V -5.04243 H -28.125\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a5 c2 node n4 p0\" transform=\"translate(-28.125 3.91764)\"><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 1.12479)\"><path class=\"edge\" d=\"M 0 0 V -1.12479 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i1 leaf node n3 p0 sample\" transform=\"translate(18.75 1.12479)\"><path class=\"edge\" d=\"M 0 0 V -1.12479 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -3.91764 H 28.125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -41.0358 H -42.1875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">7</text></g></g></g><g class=\"tree t1\" transform=\"translate(210 0)\"><g class=\"plotbox\"><g class=\"c2 node n9 p0 root\" transform=\"translate(109.062 26.8)\"><g class=\"a9 i1 leaf node n2 p0 sample\" transform=\"translate(42.1875 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H -42.1875\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a9 c2 node n7 p0\" transform=\"translate(-42.1875 48.5217)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-28.125 46.0783)\"><path class=\"edge\" d=\"M 0 0 V -46.0783 H 28.125\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 c2 node n4 p0\" transform=\"translate(28.125 44.9535)\"><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 1.12479)\"><path class=\"edge\" d=\"M 0 0 V -1.12479 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i1 leaf node n3 p0 sample\" transform=\"translate(18.75 1.12479)\"><path class=\"edge\" d=\"M 0 0 V -1.12479 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -44.9535 H -28.125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -48.5217 H 42.1875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">9</text></g></g></g><g class=\"tree t2\" transform=\"translate(400 0)\"><g class=\"plotbox\"><g class=\"c2 node n8 p0 root\" transform=\"translate(109.062 46.8017)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(42.1875 74.5983)\"><path class=\"edge\" d=\"M 0 0 V -74.5983 H -42.1875\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 c2 node n7 p0\" transform=\"translate(-42.1875 28.5201)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-28.125 46.0783)\"><path class=\"edge\" d=\"M 0 0 V -46.0783 H 28.125\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 c2 node n4 p0\" transform=\"translate(28.125 44.9535)\"><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 1.12479)\"><path class=\"edge\" d=\"M 0 0 V -1.12479 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i1 leaf node n3 p0 sample\" transform=\"translate(18.75 1.12479)\"><path class=\"edge\" d=\"M 0 0 V -1.12479 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -44.9535 H -28.125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -28.5201 H 42.1875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">8</text></g></g></g><g class=\"tree t3\" transform=\"translate(590 0)\"><g class=\"plotbox\"><g class=\"c2 node n8 p0 root\" transform=\"translate(95 46.8017)\"><g class=\"a8 c2 node n4 p0\" transform=\"translate(37.5 73.4735)\"><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 1.12479)\"><path class=\"edge\" d=\"M 0 0 V -1.12479 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i1 leaf node n3 p0 sample\" transform=\"translate(18.75 1.12479)\"><path class=\"edge\" d=\"M 0 0 V -1.12479 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -73.4735 H -37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4</text></g><g class=\"a8 c2 node n6 p0\" transform=\"translate(-37.5 35.9638)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-18.75 38.6345)\"><path class=\"edge\" d=\"M 0 0 V -38.6345 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(18.75 38.6345)\"><path class=\"edge\" d=\"M 0 0 V -38.6345 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -35.9638 H 37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">8</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=2, random_seed=28, sequence_length=100,\n",
    "    recombination_rate=0.01, discrete_genome=False)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(10.55am)\n",
    "\n",
    "*Exercise:* Our study organisms have a ploidy of 3 and a chromosome of length 100.\n",
    "Between the bases at positions 50 and 60, there is a recombination 'hotspot' with a higher recombination rate of 0.1.\n",
    "Modify the following code to simulate chromosomes for 2 such organisms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"1800\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l195.556,0 l0,138.2 l-1.95556,25 l0,5 l-193.6,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M215.556,0 l195.556,0 l0,138.2 l-91.9111,25 l0,5 l-105.6,0 l0,-5 l1.95556,-25 l0,-138.2z\"/><path d=\"M411.111,0 l195.556,0 l0,138.2 l-129.067,25 l0,5 l-158.4,0 l0,-5 l91.9111,-25 l0,-138.2z\"/><path d=\"M606.667,0 l195.556,0 l0,138.2 l-271.822,25 l0,5 l-52.8,0 l0,-5 l129.067,-25 l0,-138.2z\"/><path d=\"M802.222,0 l195.556,0 l0,138.2 l-396.978,25 l0,5 l-70.4,0 l0,-5 l271.822,-25 l0,-138.2z\"/><path d=\"M997.778,0 l195.556,0 l0,138.2 l-170.133,25 l0,5 l-422.4,0 l0,-5 l396.978,-25 l0,-138.2z\"/><path d=\"M1193.33,0 l195.556,0 l0,138.2 l-242.489,25 l0,5 l-123.2,0 l0,-5 l170.133,-25 l0,-138.2z\"/><path d=\"M1388.89,0 l195.556,0 l0,138.2 l-86.0444,25 l0,5 l-352,0 l0,-5 l242.489,-25 l0,-138.2z\"/><path d=\"M1584.44,0 l195.556,0 l0,138.2 l0,25 l0,5 l-281.6,0 l0,-5 l86.0444,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(900 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"1780\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(213.6 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">11</text></g></g><g class=\"tick\" transform=\"translate(319.2 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">17</text></g></g><g class=\"tick\" transform=\"translate(477.6 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">26</text></g></g><g class=\"tick\" transform=\"translate(530.4 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">29</text></g></g><g class=\"tick\" transform=\"translate(600.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">33</text></g></g><g class=\"tick\" transform=\"translate(1023.2 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">57</text></g></g><g class=\"tick\" transform=\"translate(1146.4 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">64</text></g></g><g class=\"tick\" transform=\"translate(1498.4 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">84</text></g></g><g class=\"tick\" transform=\"translate(1780 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">100</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n16 p0 root\" transform=\"translate(97.7778 27.9511)\"><g class=\"a16 c2 node n10 p0\" transform=\"translate(-32.4074 70.6203)\"><g class=\"a10 i1 leaf node n3 p0 sample\" transform=\"translate(19.4444 22.8286)\"><path class=\"edge\" d=\"M 0 0 V -22.8286 H -19.4444\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a10 c2 node n7 p0\" transform=\"translate(-19.4444 14.0934)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 i2 leaf node n5 p0 sample\" transform=\"translate(12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -14.0934 H 19.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -70.6203 H 32.4074\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">10</text></g><g class=\"a16 c2 node n14 p0\" transform=\"translate(32.4074 16.3959)\"><g class=\"a14 i0 leaf node n1 p0 sample\" transform=\"translate(-19.4444 77.053)\"><path class=\"edge\" d=\"M 0 0 V -77.053 H 19.4444\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a14 c2 node n6 p0\" transform=\"translate(19.4444 76.3019)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -76.3019 H -19.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -16.3959 H -32.4074\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">14</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">16</text></g></g></g><g class=\"tree t1\" transform=\"translate(215.556 0)\"><g class=\"plotbox\"><g class=\"c2 node n16 p0 root\" transform=\"translate(128.565 27.9511)\"><g class=\"a16 c2 node n14 p0\" transform=\"translate(-34.0278 16.3959)\"><g class=\"a14 c2 node n6 p0\" transform=\"translate(29.1667 76.3019)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -76.3019 H -29.1667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><g class=\"a14 c2 node n11 p0\" transform=\"translate(-29.1667 43.7712)\"><g class=\"a11 i0 leaf node n1 p0 sample\" transform=\"translate(19.4444 33.2818)\"><path class=\"edge\" d=\"M 0 0 V -33.2818 H -19.4444\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a11 c2 node n7 p0\" transform=\"translate(-19.4444 24.5466)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 i2 leaf node n5 p0 sample\" transform=\"translate(12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -24.5466 H 19.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -43.7712 H 29.1667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">11</text></g><path class=\"edge\" d=\"M 0 0 V -16.3959 H 34.0278\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">14</text></g><g class=\"a16 i1 leaf node n3 p0 sample\" transform=\"translate(34.0278 93.4489)\"><path class=\"edge\" d=\"M 0 0 V -93.4489 H -34.0278\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">16</text></g></g></g><g class=\"tree t2\" transform=\"translate(411.111 0)\"><g class=\"plotbox\"><g class=\"c2 node n16 p0 root\" transform=\"translate(104.259 27.9511)\"><g class=\"a16 c2 node n11 p0\" transform=\"translate(-38.8889 60.1671)\"><g class=\"a11 i0 leaf node n1 p0 sample\" transform=\"translate(19.4444 33.2818)\"><path class=\"edge\" d=\"M 0 0 V -33.2818 H -19.4444\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a11 c2 node n7 p0\" transform=\"translate(-19.4444 24.5466)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 i2 leaf node n5 p0 sample\" transform=\"translate(12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -24.5466 H 19.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -60.1671 H 38.8889\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">11</text></g><g class=\"a16 c2 node n15 p0\" transform=\"translate(38.8889 5.7174)\"><g class=\"a15 i1 leaf node n3 p0 sample\" transform=\"translate(19.4444 87.7315)\"><path class=\"edge\" d=\"M 0 0 V -87.7315 H -19.4444\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a15 c2 node n6 p0\" transform=\"translate(-19.4444 86.9804)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -86.9804 H 19.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -5.7174 H -38.8889\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">15</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">16</text></g></g></g><g class=\"tree t3\" transform=\"translate(606.667 0)\"><g class=\"plotbox\"><g class=\"c2 node n17 p0 root\" transform=\"translate(104.259 26.8)\"><g class=\"a17 c2 node n11 p0\" transform=\"translate(-38.8889 61.3182)\"><g class=\"a11 i0 leaf node n1 p0 sample\" transform=\"translate(19.4444 33.2818)\"><path class=\"edge\" d=\"M 0 0 V -33.2818 H -19.4444\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a11 c2 node n7 p0\" transform=\"translate(-19.4444 24.5466)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 i2 leaf node n5 p0 sample\" transform=\"translate(12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -24.5466 H 19.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -61.3182 H 38.8889\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">11</text></g><g class=\"a17 c2 node n15 p0\" transform=\"translate(38.8889 6.8685)\"><g class=\"a15 i1 leaf node n3 p0 sample\" transform=\"translate(19.4444 87.7315)\"><path class=\"edge\" d=\"M 0 0 V -87.7315 H -19.4444\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a15 c2 node n6 p0\" transform=\"translate(-19.4444 86.9804)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -86.9804 H 19.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -6.8685 H -38.8889\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">15</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">17</text></g></g></g><g class=\"tree t4\" transform=\"translate(802.222 0)\"><g class=\"plotbox\"><g class=\"c2 node n17 p0 root\" transform=\"translate(89.6759 26.8)\"><g class=\"a17 c2 node n15 p0\" transform=\"translate(43.75 6.8685)\"><g class=\"a15 i1 leaf node n3 p0 sample\" transform=\"translate(29.1667 87.7315)\"><path class=\"edge\" d=\"M 0 0 V -87.7315 H -29.1667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a15 c2 node n12 p0\" transform=\"translate(-29.1667 54.1224)\"><g class=\"a12 i0 leaf node n1 p0 sample\" transform=\"translate(-19.4444 33.6091)\"><path class=\"edge\" d=\"M 0 0 V -33.6091 H 19.4444\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a12 c2 node n6 p0\" transform=\"translate(19.4444 32.858)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -32.858 H -19.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -54.1224 H 29.1667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">12</text></g><path class=\"edge\" d=\"M 0 0 V -6.8685 H -43.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">15</text></g><g class=\"a17 c2 node n7 p0\" transform=\"translate(-43.75 85.8648)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 i2 leaf node n5 p0 sample\" transform=\"translate(12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -85.8648 H 43.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">17</text></g></g></g><g class=\"tree t5\" transform=\"translate(997.778 0)\"><g class=\"plotbox\"><g class=\"c2 node n16 p0 root\" transform=\"translate(89.6759 27.9511)\"><g class=\"a16 c2 node n7 p0\" transform=\"translate(-43.75 84.7137)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 i2 leaf node n5 p0 sample\" transform=\"translate(12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -84.7137 H 43.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><g class=\"a16 c2 node n15 p0\" transform=\"translate(43.75 5.7174)\"><g class=\"a15 i1 leaf node n3 p0 sample\" transform=\"translate(29.1667 87.7315)\"><path class=\"edge\" d=\"M 0 0 V -87.7315 H -29.1667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a15 c2 node n12 p0\" transform=\"translate(-29.1667 54.1224)\"><g class=\"a12 i0 leaf node n1 p0 sample\" transform=\"translate(-19.4444 33.6091)\"><path class=\"edge\" d=\"M 0 0 V -33.6091 H 19.4444\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a12 c2 node n6 p0\" transform=\"translate(19.4444 32.858)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -32.858 H -19.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -54.1224 H 29.1667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">12</text></g><path class=\"edge\" d=\"M 0 0 V -5.7174 H -43.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">15</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">16</text></g></g></g><g class=\"tree t6\" transform=\"translate(1193.33 0)\"><g class=\"plotbox\"><g class=\"c2 node n16 p0 root\" transform=\"translate(128.565 27.9511)\"><g class=\"a16 i1 leaf node n3 p0 sample\" transform=\"translate(34.0278 93.4489)\"><path class=\"edge\" d=\"M 0 0 V -93.4489 H -34.0278\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a16 c2 node n12 p0\" transform=\"translate(-34.0278 59.8398)\"><g class=\"a12 c2 node n6 p0\" transform=\"translate(29.1667 32.858)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -32.858 H -29.1667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><g class=\"a12 c2 node n8 p0\" transform=\"translate(-29.1667 20.617)\"><g class=\"a8 i0 leaf node n1 p0 sample\" transform=\"translate(19.4444 12.9921)\"><path class=\"edge\" d=\"M 0 0 V -12.9921 H -19.4444\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a8 c2 node n7 p0\" transform=\"translate(-19.4444 4.2569)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 i2 leaf node n5 p0 sample\" transform=\"translate(12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -4.2569 H 19.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -20.617 H 29.1667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -59.8398 H 34.0278\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">12</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">16</text></g></g></g><g class=\"tree t7\" transform=\"translate(1388.89 0)\"><g class=\"plotbox\"><g class=\"c2 node n16 p0 root\" transform=\"translate(128.565 27.9511)\"><g class=\"a16 i1 leaf node n3 p0 sample\" transform=\"translate(34.0278 93.4489)\"><path class=\"edge\" d=\"M 0 0 V -93.4489 H -34.0278\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a16 c2 node n12 p0\" transform=\"translate(-34.0278 59.8398)\"><g class=\"a12 c2 node n6 p0\" transform=\"translate(29.1667 32.858)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -32.858 H -29.1667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><g class=\"a12 c2 node n9 p0\" transform=\"translate(-29.1667 13.0888)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(19.4444 20.5203)\"><path class=\"edge\" d=\"M 0 0 V -20.5203 H -19.4444\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a9 c2 node n7 p0\" transform=\"translate(-19.4444 11.7851)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 i2 leaf node n5 p0 sample\" transform=\"translate(12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -11.7851 H 19.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -13.0888 H 29.1667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -59.8398 H 34.0278\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">12</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">16</text></g></g></g><g class=\"tree t8\" transform=\"translate(1584.44 0)\"><g class=\"plotbox\"><g class=\"c2 node n13 p0 root\" transform=\"translate(128.565 75.0283)\"><g class=\"a13 i1 leaf node n3 p0 sample\" transform=\"translate(34.0278 46.3717)\"><path class=\"edge\" d=\"M 0 0 V -46.3717 H -34.0278\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a13 c2 node n12 p0\" transform=\"translate(-34.0278 12.7626)\"><g class=\"a12 c2 node n6 p0\" transform=\"translate(29.1667 32.858)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(12.963 0.751099)\"><path class=\"edge\" d=\"M 0 0 V -0.751099 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -32.858 H -29.1667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><g class=\"a12 c2 node n9 p0\" transform=\"translate(-29.1667 13.0888)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(19.4444 20.5203)\"><path class=\"edge\" d=\"M 0 0 V -20.5203 H -19.4444\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a9 c2 node n7 p0\" transform=\"translate(-19.4444 11.7851)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H 12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 i2 leaf node n5 p0 sample\" transform=\"translate(12.963 8.73518)\"><path class=\"edge\" d=\"M 0 0 V -8.73518 H -12.963\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -11.7851 H 19.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -13.0888 H 29.1667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -12.7626 H 34.0278\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">12</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">13</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rate_map = msprime.RateMap(position=[0, 100], rate=[0.01])\n",
    "ts = msprime.sim_ancestry(samples=3,\n",
    "                          recombination_rate=rate_map,\n",
    "                          random_seed=193)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Demographic history\n",
    "\n",
    "(11am)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So far, we have been simulating samples from a single population of a constant size, which isnâ€™t particularly exciting! One of the strengths of msprime is that it can be used to specify quite complicated models of demography and population history with a simple Python API."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Population size\n",
    "\n",
    "So far, we've assumed our samples and all their ancestors belong to a single population with $2N_e=1$, which scales the simulation to use classical coalescent time units of $\\frac{1}{2N_e}$. (Note that `msprime` scales this for different ploidies, also)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n6 p0 root\" transform=\"translate(80 26.8)\"><g class=\"a6 c2 node n4 p0\" transform=\"translate(30 94.461)\"><g class=\"a4 i1 leaf node n2 p0 sample\" transform=\"translate(-15 0.139)\"><path class=\"edge\" d=\"M 0 0 V -0.139 H 15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a4 i1 leaf node n3 p0 sample\" transform=\"translate(15 0.139)\"><path class=\"edge\" d=\"M 0 0 V -0.139 H -15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -94.461 H -30\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(-30 93.1216)\"><g class=\"a5 i0 leaf node n0 p0 sample\" transform=\"translate(-15 1.47837)\"><path class=\"edge\" d=\"M 0 0 V -1.47837 H 15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a5 i0 leaf node n1 p0 sample\" transform=\"translate(15 1.47837)\"><path class=\"edge\" d=\"M 0 0 V -1.47837 H -15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><path class=\"edge\" d=\"M 0 0 V -93.1216 H 30\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(samples=2, random_seed=234)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total branch length: 14.394580159518803\n",
      "Time at root: 7.136285694725596\n"
     ]
    }
   ],
   "source": [
    "tree = ts.first()\n",
    "print(\"Total branch length:\", tree.total_branch_length)\n",
    "print(\"Time at root:\", ts.tables.nodes.time[tree.root])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By specifying a specific population size using the `population_size` argument, all times and branch lengths will be rescaled into units of generations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n6 p0 root\" transform=\"translate(80 26.8)\"><g class=\"a6 c2 node n4 p0\" transform=\"translate(30 94.461)\"><g class=\"a4 i1 leaf node n2 p0 sample\" transform=\"translate(-15 0.139)\"><path class=\"edge\" d=\"M 0 0 V -0.139 H 15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a4 i1 leaf node n3 p0 sample\" transform=\"translate(15 0.139)\"><path class=\"edge\" d=\"M 0 0 V -0.139 H -15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -94.461 H -30\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(-30 93.1216)\"><g class=\"a5 i0 leaf node n0 p0 sample\" transform=\"translate(-15 1.47837)\"><path class=\"edge\" d=\"M 0 0 V -1.47837 H 15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a5 i0 leaf node n1 p0 sample\" transform=\"translate(15 1.47837)\"><path class=\"edge\" d=\"M 0 0 V -1.47837 H -15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><path class=\"edge\" d=\"M 0 0 V -93.1216 H 30\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=2,\n",
    "    random_seed=234,\n",
    "    population_size=2)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total branch length: 28.789160319037606\n",
      "Time at root: 14.272571389451192\n"
     ]
    }
   ],
   "source": [
    "tree = ts.first()\n",
    "print(\"Total branch length:\", tree.total_branch_length)\n",
    "print(\"Time at root:\", ts.tables.nodes.time[tree.root])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that this population size should be the number of *individuals*, not the number of genomes. `msprime` uses the supplied `ploidy` to do all the other necessary adjustments under-the-hood."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total branch length: 57.57832063807521\n",
      "Time at root: 28.545142778902385\n"
     ]
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=1,\n",
    "    ploidy=4,\n",
    "    random_seed=234,\n",
    "    population_size=2)\n",
    "\n",
    "tree = ts.first()\n",
    "print(\"Total branch length:\", tree.total_branch_length)\n",
    "print(\"Time at root:\", ts.tables.nodes.time[tree.root])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Population structure\n",
    "\n",
    "(11.05am)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To simulate under more complicated models of demographic history, we will need to create a `msprime.Demography` object.\n",
    "`msprime` supports simulation from multiple discrete populations, each of which is initialized via the `add_population()` method.\n",
    "For each population, you can specify a sample size, an effective population size, an exponential growth rate and a name."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We could have specified our previous simulation with the following code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (1)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>my_pop</td><td></td><td>1.0</td><td>0</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Migration matrix (all zero)</b><table border=\"1\" class=\"tskit-table\"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events (0)</b><table border=\"1\" class=\"tskit-table\"><thead><tr></tr></thead><tbody></tbody></table></div></div>"
      ],
      "text/plain": [
       "Demography(populations=[Population(initial_size=1, growth_rate=0, name='my_pop', description='', extra_metadata={}, default_sampling_time=None, initially_active=None, id=0)], events=[], migration_matrix=array([[0.]]))"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dem = msprime.Demography()\n",
    "dem.add_population(\n",
    "    name=\"my_pop\",\n",
    "    initial_size=1\n",
    ")\n",
    "dem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n6 p0 root\" transform=\"translate(80 26.8)\"><g class=\"a6 c2 node n4 p0\" transform=\"translate(30 94.461)\"><g class=\"a4 i1 leaf node n2 p0 sample\" transform=\"translate(-15 0.139)\"><path class=\"edge\" d=\"M 0 0 V -0.139 H 15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a4 i1 leaf node n3 p0 sample\" transform=\"translate(15 0.139)\"><path class=\"edge\" d=\"M 0 0 V -0.139 H -15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -94.461 H -30\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(-30 93.1216)\"><g class=\"a5 i0 leaf node n0 p0 sample\" transform=\"translate(-15 1.47837)\"><path class=\"edge\" d=\"M 0 0 V -1.47837 H 15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a5 i0 leaf node n1 p0 sample\" transform=\"translate(15 1.47837)\"><path class=\"edge\" d=\"M 0 0 V -1.47837 H -15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><path class=\"edge\" d=\"M 0 0 V -93.1216 H 30\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=2,\n",
    "    random_seed=234,\n",
    "    demography=dem\n",
    ")\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also use `Population` objects to specify an exponential growth rate of the population per generation (i.e. forwards in time)\n",
    "\n",
    "<img src=\"pics/worksheet2-dem1.png\" width=\"400\" height=\"400\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (1)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>my_pop</td><td></td><td>100000.0</td><td>0.1</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Migration matrix (all zero)</b><table border=\"1\" class=\"tskit-table\"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events (0)</b><table border=\"1\" class=\"tskit-table\"><thead><tr></tr></thead><tbody></tbody></table></div></div>"
      ],
      "text/plain": [
       "Demography(populations=[Population(initial_size=100000.0, growth_rate=0.1, name='my_pop', description='', extra_metadata={}, default_sampling_time=None, initially_active=None, id=0)], events=[], migration_matrix=array([[0.]]))"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dem = msprime.Demography()\n",
    "dem.add_population(\n",
    "    name=\"my_pop\",\n",
    "    initial_size=1e5,\n",
    "    growth_rate=0.1\n",
    ")\n",
    "dem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n6 p0 root\" transform=\"translate(61.25 26.8)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-26.25 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 26.25\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(26.25 13.122)\"><g class=\"a5 i0 leaf node n1 p0 sample\" transform=\"translate(-22.5 81.478)\"><path class=\"edge\" d=\"M 0 0 V -81.478 H 22.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a5 c2 node n4 p0\" transform=\"translate(22.5 22.5279)\"><g class=\"a4 i1 leaf node n2 p0 sample\" transform=\"translate(-15 58.95)\"><path class=\"edge\" d=\"M 0 0 V -58.95 H 15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a4 i1 leaf node n3 p0 sample\" transform=\"translate(15 58.95)\"><path class=\"edge\" d=\"M 0 0 V -58.95 H -15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -22.5279 H -22.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -13.122 H -26.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=2,\n",
    "    random_seed=11,\n",
    "    demography=dem\n",
    ")\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that the recent branches are much longer than the older ones.\n",
    "This is what we expect to see in a growing population."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can add any number of populations into our `Demography` objects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (2)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>R</td><td>Red population</td><td>500.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>B</td><td>Blue population</td><td>500.0</td><td>0.0001</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Migration matrix (all zero)</b><table border=\"1\" class=\"tskit-table\"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events (0)</b><table border=\"1\" class=\"tskit-table\"><thead><tr></tr></thead><tbody></tbody></table></div></div>"
      ],
      "text/plain": [
       "Demography(populations=[Population(initial_size=500, growth_rate=0, name='R', description='Red population', extra_metadata={}, default_sampling_time=None, initially_active=None, id=0), Population(initial_size=500, growth_rate=0.0001, name='B', description='Blue population', extra_metadata={}, default_sampling_time=None, initially_active=None, id=1)], events=[], migration_matrix=array([[0., 0.],\n",
       "       [0., 0.]]))"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dem = msprime.Demography()\n",
    "dem.add_population(\n",
    "    name=\"R\",\n",
    "    description=\"Red population\",\n",
    "    initial_size=500\n",
    ")\n",
    "dem.add_population(\n",
    "    name=\"B\",\n",
    "    description=\"Blue population\",\n",
    "    initial_size=500,\n",
    "    growth_rate=0.0001\n",
    ")\n",
    "dem"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"pics/worksheet2-dem2.png\" width=\"400\" height=\"400\">\n",
    "\n",
    "However, this simulation will run forever unless we also specify some migration between the groups!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # without an end_time, THIS WILL RUN FOREVER!\n",
    "\n",
    "# ts = msprime.sim_ancestry(\n",
    "#   samples={\"R\" : 3, \"B\" : 3}, \n",
    "#   demography=dem,\n",
    "#   random_seed=12,\n",
    "#   sequence_length=1000, \n",
    "#   )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To understand why, recall that msprime is a backwards-time simulator. The simulation will run beginning from the present day working further back in time, simulating until all samples have coalesced to a single common ancestor at each genomic location. However, with no migration between our two populations, samples in one population will never coalesce with samples in another population. To fix this, letâ€™s add some migration events."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Migration\n",
    "\n",
    "With msprime, you can specify continual rates of migrations between populations, as well as admixture events, divergences and one-off mass migrations.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Migration rates** are passed to our `msprime.Demography` object individually via the `set_migration_rate()` method.\n",
    "This allows us to specify the expected number of migrants moving from population `dest` to population `source` per generation, divided by the size of population `source`. When this rate is small (close to 0), it is approximately equal to the fraction of population `source` that consists of new migrants from population `dest` in each generation. \n",
    "\n",
    "If both of your populations are exchanging migrants at the same rate, you can save yourself some typing by specifying them with a single `set_symmetric_migration_rate()` call.\n",
    "See the documentation [here](https://tskit.dev/msprime/docs/stable/api.html#msprime.Demography.add_symmetric_migration_rate_change)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "dem = msprime.Demography()\n",
    "dem.add_population(\n",
    "    name=\"R\",\n",
    "    description=\"Red population\",\n",
    "    initial_size=500\n",
    ")\n",
    "dem.add_population(\n",
    "    name=\"B\",\n",
    "    description=\"Blue population\",\n",
    "    initial_size=500,\n",
    "    growth_rate=0.0001\n",
    ")\n",
    "dem\n",
    "\n",
    "# Set migration rates.\n",
    "dem.set_migration_rate(source=0, dest=1, rate=0.05)\n",
    "dem.set_migration_rate(source=1, dest=0, rate=0.02)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For instance, the `Demography` object above specifies that in each generation, approximately 5% of the red population (0) consists of migrants from the blue population (1), and approximately 2% of the blue population consists of migrants from the red population."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "              <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "              </style>\n",
       "              <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"padding:0;line-height:21px;\">\n",
       "                          <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                          <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                        </th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                      <tr><td>Trees</td><td>1</td></tr>\n",
       "                      <tr><td>Sequence Length</td><td>10000000.0</td></tr>\n",
       "                      <tr><td>Time Units</td><td>generations</td></tr>\n",
       "                      <tr><td>Sample Nodes</td><td>20</td></tr>\n",
       "                      <tr><td>Total Size</td><td>4.6 KiB</td></tr>\n",
       "                      <tr>\n",
       "                        <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"line-height:21px;\">Table</th>\n",
       "                        <th>Rows</th>\n",
       "                        <th>Size</th>\n",
       "                        <th>Has Metadata</th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                    \n",
       "                  <tr>\n",
       "                    <td>Edges</td>\n",
       "                      <td>38</td>\n",
       "                      <td>1.2 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Individuals</td>\n",
       "                      <td>10</td>\n",
       "                      <td>304 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Migrations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>8 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Mutations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>16 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Nodes</td>\n",
       "                      <td>39</td>\n",
       "                      <td>1.1 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Populations</td>\n",
       "                      <td>2</td>\n",
       "                      <td>286 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        âœ…\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Provenances</td>\n",
       "                      <td>1</td>\n",
       "                      <td>1.4 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Sites</td>\n",
       "                      <td>0</td>\n",
       "                      <td>16 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "              </div>\n",
       "            </div>\n",
       "            "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x1385f0eb8>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Simulate.\n",
    "ts = msprime.sim_ancestry(\n",
    "  samples={\"R\" : 5, \"B\" : 5},\n",
    "  demography=dem,\n",
    "  sequence_length=1e7,\n",
    "  random_seed=145)\n",
    "ts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One consequence of explicitly specifying `Population` objects is that each of the simulated nodes will now belong to one of our specified populations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "                <style scoped=\"\">\n",
       "                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}\n",
       "                    .tskit-table tbody tr th {vertical-align: top;}\n",
       "                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "                    .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "                </style>\n",
       "                <table border=\"1\" class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                        <tr>\n",
       "                            <th>id</th><th>flags</th><th>population</th><th>individual</th><th>time</th><th>metadata</th>\n",
       "                        </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                        <tr><td>0</td><td>1</td><td>0</td><td>0</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>1</td><td>1</td><td>0</td><td>0</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>2</td><td>1</td><td>0</td><td>1</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>3</td><td>1</td><td>0</td><td>1</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>4</td><td>1</td><td>0</td><td>2</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>5</td><td>1</td><td>0</td><td>2</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>6</td><td>1</td><td>0</td><td>3</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>7</td><td>1</td><td>0</td><td>3</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>8</td><td>1</td><td>0</td><td>4</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>9</td><td>1</td><td>0</td><td>4</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>10</td><td>1</td><td>1</td><td>5</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>11</td><td>1</td><td>1</td><td>5</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>12</td><td>1</td><td>1</td><td>6</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>13</td><td>1</td><td>1</td><td>6</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>14</td><td>1</td><td>1</td><td>7</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>15</td><td>1</td><td>1</td><td>7</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>16</td><td>1</td><td>1</td><td>8</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>17</td><td>1</td><td>1</td><td>8</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>18</td><td>1</td><td>1</td><td>9</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>19</td><td>1</td><td>1</td><td>9</td><td>0.00000000</td><td>b''</td></tr>\n",
       "<tr><td>20</td><td>0</td><td>1</td><td>-1</td><td>13.99211400</td><td>b''</td></tr>\n",
       "<tr><td>21</td><td>0</td><td>1</td><td>-1</td><td>26.92875440</td><td>b''</td></tr>\n",
       "<tr><td>22</td><td>0</td><td>1</td><td>-1</td><td>27.31527904</td><td>b''</td></tr>\n",
       "<tr><td>23</td><td>0</td><td>1</td><td>-1</td><td>39.20615233</td><td>b''</td></tr>\n",
       "<tr><td>24</td><td>0</td><td>1</td><td>-1</td><td>40.66830357</td><td>b''</td></tr>\n",
       "<tr><td>25</td><td>0</td><td>1</td><td>-1</td><td>45.04458357</td><td>b''</td></tr>\n",
       "<tr><td>26</td><td>0</td><td>1</td><td>-1</td><td>65.82543141</td><td>b''</td></tr>\n",
       "<tr><td>27</td><td>0</td><td>0</td><td>-1</td><td>71.72151731</td><td>b''</td></tr>\n",
       "<tr><td>28</td><td>0</td><td>1</td><td>-1</td><td>82.26627457</td><td>b''</td></tr>\n",
       "<tr><td>29</td><td>0</td><td>1</td><td>-1</td><td>114.95071876</td><td>b''</td></tr>\n",
       "<tr><td>30</td><td>0</td><td>1</td><td>-1</td><td>122.30159434</td><td>b''</td></tr>\n",
       "<tr><td>31</td><td>0</td><td>1</td><td>-1</td><td>134.25571820</td><td>b''</td></tr>\n",
       "<tr><td>32</td><td>0</td><td>0</td><td>-1</td><td>234.79024032</td><td>b''</td></tr>\n",
       "<tr><td>33</td><td>0</td><td>1</td><td>-1</td><td>535.10649119</td><td>b''</td></tr>\n",
       "<tr><td>34</td><td>0</td><td>0</td><td>-1</td><td>645.64562624</td><td>b''</td></tr>\n",
       "<tr><td>35</td><td>0</td><td>1</td><td>-1</td><td>993.97961286</td><td>b''</td></tr>\n",
       "<tr><td>36</td><td>0</td><td>1</td><td>-1</td><td>2044.67749220</td><td>b''</td></tr>\n",
       "<tr><td>37</td><td>0</td><td>1</td><td>-1</td><td>3169.91050628</td><td>b''</td></tr>\n",
       "<tr><td>38</td><td>0</td><td>1</td><td>-1</td><td>3479.86504149</td><td>b''</td></tr>\n",
       "\n",
       "                    </tbody>\n",
       "                </table>\n",
       "            </div>\n",
       "        "
      ],
      "text/plain": [
       "<tskit.tables.NodeTable at 0x1385e2a20>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts.tables.nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "                <style scoped=\"\">\n",
       "                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}\n",
       "                    .tskit-table tbody tr th {vertical-align: top;}\n",
       "                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "                    .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "                </style>\n",
       "                <table border=\"1\" class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                        <tr>\n",
       "                            <th>id</th><th>metadata</th>\n",
       "                        </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                        <tr><td>0</td><td>{'description': 'Red population', 'name': 'R'}</td></tr>\n",
       "<tr><td>1</td><td>{'description': 'Blue population', 'name': 'B'}</td></tr>\n",
       "\n",
       "                    </tbody>\n",
       "                </table>\n",
       "            </div>\n",
       "        "
      ],
      "text/plain": [
       "<tskit.tables.PopulationTable at 0x1385e2208>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts.tables.populations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll use this information to draw the tree sequence with nodes coloured by population label using SVG:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tree on interval: Interval(left=0.0, right=10000000.0)\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"400\" version=\"1.1\" width=\"500\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree t0\"><g class=\"plotbox\"><g class=\"c2 node n38 p1 root\" transform=\"translate(304.805 26.8)\"><g class=\"a38 c2 node n34 p0\" transform=\"translate(98.2891 278.057)\"><g class=\"a34 c2 node n31 p1\" transform=\"translate(53.9062 50.1711)\"><g class=\"a31 i7 leaf node n15 p1 sample\" transform=\"translate(-11.5 13.1715)\"><path class=\"edge\" d=\"M 0 0 V -13.1715 H 11.5\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">15</text></g><g class=\"a31 i9 leaf node n18 p1 sample\" transform=\"translate(11.5 13.1715)\"><path class=\"edge\" d=\"M 0 0 V -13.1715 H -11.5\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">18</text></g><path class=\"edge\" d=\"M 0 0 V -50.1711 H -53.9062\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">31</text></g><g class=\"a34 c2 node n33 p1\" transform=\"translate(-53.9062 10.8447)\"><g class=\"a33 c2 node n23 p1\" transform=\"translate(30.1875 48.6514)\"><g class=\"a23 i2 leaf node n4 p0 sample\" transform=\"translate(-25.875 3.84641)\"><path class=\"edge\" d=\"M 0 0 V -3.84641 H 25.875\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a23 c2 node n22 p1\" transform=\"translate(25.875 1.16658)\"><g class=\"a22 i3 leaf node n7 p0 sample\" transform=\"translate(17.25 2.67983)\"><path class=\"edge\" d=\"M 0 0 V -2.67983 H -17.25\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">7</text></g><g class=\"a22 c2 node n21 p1\" transform=\"translate(-17.25 0.0379209)\"><g class=\"a21 i3 leaf node n6 p0 sample\" transform=\"translate(-11.5 2.64191)\"><path class=\"edge\" d=\"M 0 0 V -2.64191 H 11.5\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">6</text></g><g class=\"a21 i5 leaf node n11 p1 sample\" transform=\"translate(11.5 2.64191)\"><path class=\"edge\" d=\"M 0 0 V -2.64191 H -11.5\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">11</text></g><path class=\"edge\" d=\"M 0 0 V -0.0379209 H 17.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">21</text></g><path class=\"edge\" d=\"M 0 0 V -1.16658 H -25.875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">22</text></g><path class=\"edge\" d=\"M 0 0 V -48.6514 H -30.1875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">23</text></g><g class=\"a33 c2 node n32 p0\" transform=\"translate(-30.1875 29.4632)\"><g class=\"a32 i0 leaf node n1 p0 sample\" transform=\"translate(-11.5 23.0346)\"><path class=\"edge\" d=\"M 0 0 V -23.0346 H 11.5\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a32 i9 leaf node n19 p1 sample\" transform=\"translate(11.5 23.0346)\"><path class=\"edge\" d=\"M 0 0 V -23.0346 H -11.5\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">19</text></g><path class=\"edge\" d=\"M 0 0 V -29.4632 H 30.1875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">32</text></g><path class=\"edge\" d=\"M 0 0 V -10.8447 H 53.9062\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">33</text></g><path class=\"edge\" d=\"M 0 0 V -278.057 H -98.2891\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">34</text></g><g class=\"a38 c2 node n37 p1\" transform=\"translate(-98.2891 30.4088)\"><g class=\"a37 c2 node n29 p1\" transform=\"translate(66.4844 299.714)\"><g class=\"a29 i1 leaf node n3 p0 sample\" transform=\"translate(-11.5 11.2775)\"><path class=\"edge\" d=\"M 0 0 V -11.2775 H 11.5\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a29 i2 leaf node n5 p0 sample\" transform=\"translate(11.5 11.2775)\"><path class=\"edge\" d=\"M 0 0 V -11.2775 H -11.5\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -299.714 H -66.4844\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">29</text></g><g class=\"a37 c2 node n36 p1\" transform=\"translate(-66.4844 110.394)\"><g class=\"a36 c2 node n30 p1\" transform=\"translate(65.4062 188.599)\"><g class=\"a30 i4 leaf node n8 p0 sample\" transform=\"translate(33.0625 11.9987)\"><path class=\"edge\" d=\"M 0 0 V -11.9987 H -33.0625\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">8</text></g><g class=\"a30 c2 node n28 p1\" transform=\"translate(-33.0625 3.92776)\"><g class=\"a28 c2 node n20 p1\" transform=\"translate(31.625 6.69819)\"><g class=\"a20 i4 leaf node n9 p0 sample\" transform=\"translate(-11.5 1.37273)\"><path class=\"edge\" d=\"M 0 0 V -1.37273 H 11.5\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">9</text></g><g class=\"a20 i6 leaf node n12 p1 sample\" transform=\"translate(11.5 1.37273)\"><path class=\"edge\" d=\"M 0 0 V -1.37273 H -11.5\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">12</text></g><path class=\"edge\" d=\"M 0 0 V -6.69819 H -31.625\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">20</text></g><g class=\"a28 c2 node n26 p1\" transform=\"translate(-31.625 1.61297)\"><g class=\"a26 i1 leaf node n2 p0 sample\" transform=\"translate(-17.25 6.45795)\"><path class=\"edge\" d=\"M 0 0 V -6.45795 H 17.25\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a26 c2 node n25 p1\" transform=\"translate(17.25 2.03875)\"><g class=\"a25 i5 leaf node n10 p1 sample\" transform=\"translate(-11.5 4.4192)\"><path class=\"edge\" d=\"M 0 0 V -4.4192 H 11.5\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">10</text></g><g class=\"a25 i7 leaf node n14 p1 sample\" transform=\"translate(11.5 4.4192)\"><path class=\"edge\" d=\"M 0 0 V -4.4192 H -11.5\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">14</text></g><path class=\"edge\" d=\"M 0 0 V -2.03875 H -17.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">25</text></g><path class=\"edge\" d=\"M 0 0 V -1.61297 H 31.625\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">26</text></g><path class=\"edge\" d=\"M 0 0 V -3.92776 H 33.0625\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">28</text></g><path class=\"edge\" d=\"M 0 0 V -188.599 H -65.4062\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">30</text></g><g class=\"a36 c2 node n35 p1\" transform=\"translate(-65.4062 103.081)\"><g class=\"a35 i8 leaf node n17 p1 sample\" transform=\"translate(25.875 97.5166)\"><path class=\"edge\" d=\"M 0 0 V -97.5166 H -25.875\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">17</text></g><g class=\"a35 c2 node n27 p0\" transform=\"translate(-25.875 90.4802)\"><g class=\"a27 i0 leaf node n0 p0 sample\" transform=\"translate(-17.25 7.0364)\"><path class=\"edge\" d=\"M 0 0 V -7.0364 H 17.25\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a27 c2 node n24 p1\" transform=\"translate(17.25 3.04655)\"><g class=\"a24 i6 leaf node n13 p1 sample\" transform=\"translate(-11.5 3.98986)\"><path class=\"edge\" d=\"M 0 0 V -3.98986 H 11.5\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">13</text></g><g class=\"a24 i8 leaf node n16 p1 sample\" transform=\"translate(11.5 3.98986)\"><path class=\"edge\" d=\"M 0 0 V -3.98986 H -11.5\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">16</text></g><path class=\"edge\" d=\"M 0 0 V -3.04655 H -17.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">24</text></g><path class=\"edge\" d=\"M 0 0 V -90.4802 H 25.875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">27</text></g><path class=\"edge\" d=\"M 0 0 V -103.081 H 65.4062\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">35</text></g><path class=\"edge\" d=\"M 0 0 V -110.394 H 66.4844\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">36</text></g><path class=\"edge\" d=\"M 0 0 V -30.4088 H 98.2891\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">37</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab\" transform=\"translate(0 -11)\">38</text></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colour_map = {0:\"red\", 1:\"blue\"}\n",
    "node_colours = {u.id: colour_map[u.population] for u in ts.nodes()}\n",
    "for tree in ts.trees():\n",
    "    print(\"Tree on interval:\", tree.interval)\n",
    "    # The code below will only work in a Jupyter notebook with SVG output enabled.\n",
    "    display(SVG(tree.draw(node_colours=node_colours, width=500, height=400)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "More coalescences are happening in the blue population (1) than the red population (0). This makes sense given that population 1 is specifying more migrants to population 0 than vice versa."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Changing migration rates\n",
    "\n",
    "(11.15am)\n",
    "\n",
    "We can change any of the migration rates at any time in the simulation. To do this, we'll use the `add_migration_rate_change()` method on our demography object.\n",
    "This will specify the populations whose migration rates are to be changed, the time of the change and the new migration rate.\n",
    "\n",
    "For instance, say we wanted to specify that in each generation prior to `time=100`, 1% of the red population (0) consisted of migrants from the blue population (1)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (2)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>R</td><td>Red population</td><td>500.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>B</td><td>Blue population</td><td>500.0</td><td>0.0001</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Migration matrix</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>R</th><th>B</th></tr></thead><tbody><tr><td>R</td><td>0</td><td title='Backwards in time migration rate from population R to B = 0.05 per generation.\n",
       "Forwards in time, this is the expected number of migrants moving from B to R per\n",
       "generation, divided by the size of R.'>0.05</td></tr><tr><td>B</td><td title='Backwards in time migration rate from population B to R = 0.02 per generation.\n",
       "Forwards in time, this is the expected number of migrants moving from R to B per\n",
       "generation, divided by the size of B.'>0.02</td><td>0</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events (1)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>100</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_migration_rate_change'>Migration rate change</a></td><td>source=0, dest=1, rate=0.01</td><td>Backwards-time migration rate from 0 to 1 â†’ 0.01</td></tr></tbody></table></div></div>"
      ],
      "text/plain": [
       "Demography(populations=[Population(initial_size=500, growth_rate=0, name='R', description='Red population', extra_metadata={}, default_sampling_time=None, initially_active=None, id=0), Population(initial_size=500, growth_rate=0.0001, name='B', description='Blue population', extra_metadata={}, default_sampling_time=None, initially_active=None, id=1)], events=[MigrationRateChange(time=100, rate=0.01, source=0, dest=1)], migration_matrix=array([[0.  , 0.05],\n",
       "       [0.02, 0.  ]]))"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dem.add_migration_rate_change(time=100, rate=0.01, source=0, dest=1)\n",
    "dem"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output above shows that we have successfully added our first demographic event to our `Demography` object, a migration rate change. We are now ready to simulate:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "              <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "              </style>\n",
       "              <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"padding:0;line-height:21px;\">\n",
       "                          <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                          <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                        </th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                      <tr><td>Trees</td><td>1</td></tr>\n",
       "                      <tr><td>Sequence Length</td><td>1000.0</td></tr>\n",
       "                      <tr><td>Time Units</td><td>generations</td></tr>\n",
       "                      <tr><td>Sample Nodes</td><td>16</td></tr>\n",
       "                      <tr><td>Total Size</td><td>4.1 KiB</td></tr>\n",
       "                      <tr>\n",
       "                        <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"line-height:21px;\">Table</th>\n",
       "                        <th>Rows</th>\n",
       "                        <th>Size</th>\n",
       "                        <th>Has Metadata</th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                    \n",
       "                  <tr>\n",
       "                    <td>Edges</td>\n",
       "                      <td>30</td>\n",
       "                      <td>968 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Individuals</td>\n",
       "                      <td>8</td>\n",
       "                      <td>248 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Migrations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>8 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Mutations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>16 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Nodes</td>\n",
       "                      <td>31</td>\n",
       "                      <td>876 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Populations</td>\n",
       "                      <td>2</td>\n",
       "                      <td>286 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        âœ…\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Provenances</td>\n",
       "                      <td>1</td>\n",
       "                      <td>1.5 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Sites</td>\n",
       "                      <td>0</td>\n",
       "                      <td>16 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "              </div>\n",
       "            </div>\n",
       "            "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x1385e2828>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "  samples={\"R\" : 4, \"B\" : 4},\n",
    "  demography=dem,\n",
    "  sequence_length=1000,\n",
    "  random_seed=63461)\n",
    "ts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tree on interval: Interval(left=0.0, right=1000.0)\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"400\" version=\"1.1\" width=\"500\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree t0\"><g class=\"plotbox\"><g class=\"c2 node n30 p0 root\" transform=\"translate(185.593 26.8)\"><g class=\"a30 i0 leaf node n0 p0 sample\" transform=\"translate(-151.218 341.4)\"><path class=\"edge\" d=\"M 0 0 V -341.4 H 151.218\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a30 c2 node n29 p0\" transform=\"translate(151.218 216.019)\"><g class=\"a29 c2 node n27 p0\" transform=\"translate(-114.438 54.8603)\"><g class=\"a27 c2 node n19 p1\" transform=\"translate(-123.311 66.5773)\"><g class=\"a19 i5 leaf node n11 p1 sample\" transform=\"translate(21.5625 3.94382)\"><path class=\"edge\" d=\"M 0 0 V -3.94382 H -21.5625\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">11</text></g><g class=\"a19 c2 node n18 p1\" transform=\"translate(-21.5625 0.401329)\"><g class=\"a18 i0 leaf node n1 p0 sample\" transform=\"translate(-14.375 3.54249)\"><path class=\"edge\" d=\"M 0 0 V -3.54249 H 14.375\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a18 i6 leaf node n12 p1 sample\" transform=\"translate(14.375 3.54249)\"><path class=\"edge\" d=\"M 0 0 V -3.54249 H -14.375\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">12</text></g><path class=\"edge\" d=\"M 0 0 V -0.401329 H 21.5625\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">18</text></g><path class=\"edge\" d=\"M 0 0 V -66.5773 H 123.311\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">19</text></g><g class=\"a27 c2 node n26 p0\" transform=\"translate(123.311 17.1831)\"><g class=\"a26 c2 node n22 p1\" transform=\"translate(48.0664 43.8439)\"><g class=\"a22 i3 leaf node n7 p0 sample\" transform=\"translate(-14.375 9.49417)\"><path class=\"edge\" d=\"M 0 0 V -9.49417 H 14.375\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">7</text></g><g class=\"a22 i5 leaf node n10 p1 sample\" transform=\"translate(14.375 9.49417)\"><path class=\"edge\" d=\"M 0 0 V -9.49417 H -14.375\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">10</text></g><path class=\"edge\" d=\"M 0 0 V -43.8439 H -48.0664\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">22</text></g><g class=\"a26 c2 node n25 p0\" transform=\"translate(-48.0664 34.3188)\"><g class=\"a25 i7 leaf node n14 p1 sample\" transform=\"translate(53.0078 19.0192)\"><path class=\"edge\" d=\"M 0 0 V -19.0192 H -53.0078\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">14</text></g><g class=\"a25 c2 node n24 p1\" transform=\"translate(-53.0078 6.83133)\"><g class=\"a24 c2 node n21 p1\" transform=\"translate(41.3281 6.70682)\"><g class=\"a21 i2 leaf node n4 p0 sample\" transform=\"translate(-21.5625 5.48104)\"><path class=\"edge\" d=\"M 0 0 V -5.48104 H 21.5625\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a21 c2 node n16 p0\" transform=\"translate(21.5625 5.03893)\"><g class=\"a16 i2 leaf node n5 p0 sample\" transform=\"translate(-14.375 0.442109)\"><path class=\"edge\" d=\"M 0 0 V -0.442109 H 14.375\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><g class=\"a16 i6 leaf node n13 p1 sample\" transform=\"translate(14.375 0.442109)\"><path class=\"edge\" d=\"M 0 0 V -0.442109 H -14.375\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">13</text></g><path class=\"edge\" d=\"M 0 0 V -5.03893 H -21.5625\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">16</text></g><path class=\"edge\" d=\"M 0 0 V -6.70682 H -41.3281\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">21</text></g><g class=\"a24 c2 node n23 p1\" transform=\"translate(-41.3281 0.850516)\"><g class=\"a23 i7 leaf node n15 p1 sample\" transform=\"translate(32.3438 11.3373)\"><path class=\"edge\" d=\"M 0 0 V -11.3373 H -32.3438\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">15</text></g><g class=\"a23 c2 node n20 p1\" transform=\"translate(-32.3438 7.06054)\"><g class=\"a20 i1 leaf node n2 p0 sample\" transform=\"translate(-21.5625 4.2768)\"><path class=\"edge\" d=\"M 0 0 V -4.2768 H 21.5625\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a20 c2 node n17 p1\" transform=\"translate(21.5625 1.45557)\"><g class=\"a17 i4 leaf node n8 p1 sample\" transform=\"translate(-14.375 2.82123)\"><path class=\"edge\" d=\"M 0 0 V -2.82123 H 14.375\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">8</text></g><g class=\"a17 i4 leaf node n9 p1 sample\" transform=\"translate(14.375 2.82123)\"><path class=\"edge\" d=\"M 0 0 V -2.82123 H -14.375\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -1.45557 H -21.5625\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">17</text></g><path class=\"edge\" d=\"M 0 0 V -7.06054 H 32.3438\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">20</text></g><path class=\"edge\" d=\"M 0 0 V -0.850516 H 41.3281\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">23</text></g><path class=\"edge\" d=\"M 0 0 V -6.83133 H 53.0078\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">24</text></g><path class=\"edge\" d=\"M 0 0 V -34.3188 H 48.0664\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">25</text></g><path class=\"edge\" d=\"M 0 0 V -17.1831 H -123.311\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">26</text></g><path class=\"edge\" d=\"M 0 0 V -54.8603 H 114.438\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">27</text></g><g class=\"a29 c2 node n28 p0\" transform=\"translate(114.438 5.79343)\"><g class=\"a28 i1 leaf node n3 p0 sample\" transform=\"translate(-14.375 119.588)\"><path class=\"edge\" d=\"M 0 0 V -119.588 H 14.375\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a28 i3 leaf node n6 p0 sample\" transform=\"translate(14.375 119.588)\"><path class=\"edge\" d=\"M 0 0 V -119.588 H -14.375\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -5.79343 H -114.438\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">28</text></g><path class=\"edge\" d=\"M 0 0 V -216.019 H -151.218\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">29</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab\" transform=\"translate(0 -11)\">30</text></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colour_map = {0:\"red\", 1:\"blue\"}\n",
    "node_colours = {u.id: colour_map[u.population] for u in ts.nodes()}\n",
    "for tree in ts.trees():\n",
    "    print(\"Tree on interval:\", tree.interval)\n",
    "    # The code below will only work in a Jupyter notebook with SVG output enabled.\n",
    "    display(SVG(tree.draw(node_colours=node_colours, width=500, height=400)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Admixture\n",
    "\n",
    "(11.25am)\n",
    "\n",
    "It is also easy to specify admixture and divergence events with `msprime`. Suppose we wanted to specify our demography so that our admixed population arose 200 generations ago, with 70% of the new population being migrants from the red ancestral population, and 30% being migrants from the blue ancestral population.\n",
    "\n",
    "<img src=\"pics/worksheet2-dem3.png\" width=\"400\" height=\"400\">\n",
    "\n",
    "We can do this by using the `add_admixture()` method on our demography object. We must supply a list of ancestral populations participating in the admixture, and a list of the same size specifying the proportions of migrants from each of these populations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (3)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>AncestralPop0</td><td>Plotted in red.</td><td>1000.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>AncestralPop1</td><td>Plotted in blue.</td><td>1000.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>AdmixedPop</td><td>Plotted in purple.</td><td>700.0</td><td>0</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Migration matrix</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>AncestralPop0</th><th>AncestralPop1</th><th>AdmixedPop</th></tr></thead><tbody><tr><td>AncestralPop0</td><td>0</td><td title='Backwards in time migration rate from population AncestralPop0 to AncestralPop1\n",
       "= 0.001 per generation. Forwards in time, this is the expected number of\n",
       "migrants moving from AncestralPop1 to AncestralPop0 per generation, divided by\n",
       "the size of AncestralPop0.'>0.001</td><td title='Backwards in time migration rate from population AncestralPop0 to AdmixedPop =\n",
       "0.0 per generation. Forwards in time, this is the expected number of migrants\n",
       "moving from AdmixedPop to AncestralPop0 per generation, divided by the size of\n",
       "AncestralPop0.'>0</td></tr><tr><td>AncestralPop1</td><td title='Backwards in time migration rate from population AncestralPop1 to AncestralPop0\n",
       "= 0.001 per generation. Forwards in time, this is the expected number of\n",
       "migrants moving from AncestralPop0 to AncestralPop1 per generation, divided by\n",
       "the size of AncestralPop1.'>0.001</td><td>0</td><td title='Backwards in time migration rate from population AncestralPop1 to AdmixedPop =\n",
       "0.0 per generation. Forwards in time, this is the expected number of migrants\n",
       "moving from AdmixedPop to AncestralPop1 per generation, divided by the size of\n",
       "AncestralPop1.'>0</td></tr><tr><td>AdmixedPop</td><td title='Backwards in time migration rate from population AdmixedPop to AncestralPop0 =\n",
       "0.0 per generation. Forwards in time, this is the expected number of migrants\n",
       "moving from AncestralPop0 to AdmixedPop per generation, divided by the size of\n",
       "AdmixedPop.'>0</td><td title='Backwards in time migration rate from population AdmixedPop to AncestralPop1 =\n",
       "0.0 per generation. Forwards in time, this is the expected number of migrants\n",
       "moving from AncestralPop1 to AdmixedPop per generation, divided by the size of\n",
       "AdmixedPop.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events (1)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>200</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_admixture'>Admixture</a></td><td>derived=AdmixedPop ancestral=[AncestralPop0, AncestralPop1] proportions=[0.60, 0.40]</td><td>Moves all lineages from admixed population &#x27;AdmixedPop&#x27; to ancestral populations. Lineages move to &#x27;AncestralPop0&#x27; with proba 0.6; &#x27;AncestralPop1&#x27; with proba 0.4. Set &#x27;AdmixedPop&#x27; to inactive, and all migration rates to and from &#x27;AdmixedPop&#x27; to zero.</td></tr></tbody></table></div></div>"
      ],
      "text/plain": [
       "Demography(populations=[Population(initial_size=1000, growth_rate=0, name='AncestralPop0', description='Plotted in red.', extra_metadata={}, default_sampling_time=None, initially_active=None, id=0), Population(initial_size=1000, growth_rate=0, name='AncestralPop1', description='Plotted in blue.', extra_metadata={}, default_sampling_time=None, initially_active=None, id=1), Population(initial_size=700, growth_rate=0, name='AdmixedPop', description='Plotted in purple.', extra_metadata={}, default_sampling_time=None, initially_active=None, id=2)], events=[Admixture(time=200, derived='AdmixedPop', ancestral=['AncestralPop0', 'AncestralPop1'], proportions=[0.6, 0.4])], migration_matrix=array([[0.   , 0.001, 0.   ],\n",
       "       [0.001, 0.   , 0.   ],\n",
       "       [0.   , 0.   , 0.   ]]))"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dem = msprime.Demography()\n",
    "dem.add_population(\n",
    "    name=\"AncestralPop0\", description=\"Plotted in red.\", initial_size=1000, growth_rate=0)\n",
    "dem.add_population(\n",
    "    name=\"AncestralPop1\", description=\"Plotted in blue.\", initial_size=1000, growth_rate=0)\n",
    "dem.add_population(\n",
    "    name=\"AdmixedPop\", description=\"Plotted in purple.\", initial_size=700)\n",
    "dem.set_migration_rate(\n",
    "    source=0, dest=1, rate=0.001)\n",
    "dem.set_migration_rate(\n",
    "    source=1, dest=0, rate=0.001)\n",
    "\n",
    "# Specify admixture event.\n",
    "dem.add_admixture(time=200, derived=\"AdmixedPop\", ancestral=[\"AncestralPop0\", \"AncestralPop1\"], proportions=[0.6, 0.4])\n",
    "dem"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this simulated sample, all of the recent ancestral haplotypes belong to the admixed population, and all of those prior to the admixture event belong to one of the ancestral populations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populations of nodes from time < 500:\n",
      "[2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]\n",
      "Populations of nodes from time >= 500:\n",
      "[0, 1, 1, 0, 0]\n"
     ]
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "  samples={\"AncestralPop0\" : 0, \"AncestralPop1\" : 0, \"AdmixedPop\" : 6},\n",
    "  demography=dem,\n",
    "  sequence_length=1000,\n",
    "  random_seed=63\n",
    ")\n",
    "\n",
    "print(\"Populations of nodes from time < 500:\")\n",
    "print([u.population for u in ts.nodes() if u.time < 500])\n",
    "print(\"Populations of nodes from time >= 500:\")\n",
    "print([u.population for u in ts.nodes() if u.time >= 500])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tree on interval: Interval(left=0.0, right=1000.0)\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"400\" version=\"1.1\" width=\"500\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree t0\"><g class=\"plotbox\"><g class=\"c2 node n22 p0 root\" transform=\"translate(360.208 26.8)\"><g class=\"a22 c2 node n14 p2\" transform=\"translate(81.4583 339.802)\"><g class=\"a14 i0 leaf node n1 p2 sample\" transform=\"translate(-19.1667 1.59835)\"><path class=\"edge\" d=\"M 0 0 V -1.59835 H 19.1667\"/><rect class=\"sym\" height=\"6\" style=\"fill:purple;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a14 i4 leaf node n9 p2 sample\" transform=\"translate(19.1667 1.59835)\"><path class=\"edge\" d=\"M 0 0 V -1.59835 H -19.1667\"/><rect class=\"sym\" height=\"6\" style=\"fill:purple;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -339.802 H -81.4583\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:purple;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">14</text></g><g class=\"a22 c2 node n21 p0\" transform=\"translate(-81.4583 8.31238)\"><g class=\"a21 c2 node n16 p2\" transform=\"translate(86.25 328.814)\"><g class=\"a16 i2 leaf node n4 p2 sample\" transform=\"translate(-19.1667 4.27315)\"><path class=\"edge\" d=\"M 0 0 V -4.27315 H 19.1667\"/><rect class=\"sym\" height=\"6\" style=\"fill:purple;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a16 i4 leaf node n8 p2 sample\" transform=\"translate(19.1667 4.27315)\"><path class=\"edge\" d=\"M 0 0 V -4.27315 H -19.1667\"/><rect class=\"sym\" height=\"6\" style=\"fill:purple;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -328.814 H -86.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:purple;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">16</text></g><g class=\"a21 c2 node n20 p1\" transform=\"translate(-86.25 271.086)\"><g class=\"a20 c2 node n17 p2\" transform=\"translate(-81.4583 54.1937)\"><g class=\"a17 i5 leaf node n11 p2 sample\" transform=\"translate(43.125 7.80824)\"><path class=\"edge\" d=\"M 0 0 V -7.80824 H -43.125\"/><rect class=\"sym\" height=\"6\" style=\"fill:purple;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">11</text></g><g class=\"a17 c2 node n13 p2\" transform=\"translate(-43.125 7.11917)\"><g class=\"a13 i0 leaf node n0 p2 sample\" transform=\"translate(-28.75 0.689075)\"><path class=\"edge\" d=\"M 0 0 V -0.689075 H 28.75\"/><rect class=\"sym\" height=\"6\" style=\"fill:purple;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a13 c2 node n12 p2\" transform=\"translate(28.75 0.0180362)\"><g class=\"a12 i3 leaf node n6 p2 sample\" transform=\"translate(-19.1667 0.671039)\"><path class=\"edge\" d=\"M 0 0 V -0.671039 H 19.1667\"/><rect class=\"sym\" height=\"6\" style=\"fill:purple;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">6</text></g><g class=\"a12 i3 leaf node n7 p2 sample\" transform=\"translate(19.1667 0.671039)\"><path class=\"edge\" d=\"M 0 0 V -0.671039 H -19.1667\"/><rect class=\"sym\" height=\"6\" style=\"fill:purple;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -0.0180362 H -28.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:purple;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">12</text></g><path class=\"edge\" d=\"M 0 0 V -7.11917 H 43.125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:purple;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">13</text></g><path class=\"edge\" d=\"M 0 0 V -54.1937 H 81.4583\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:purple;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">17</text></g><g class=\"a20 c2 node n19 p1\" transform=\"translate(81.4583 21.7707)\"><g class=\"a19 i1 leaf node n3 p2 sample\" transform=\"translate(33.5417 40.2312)\"><path class=\"edge\" d=\"M 0 0 V -40.2312 H -33.5417\"/><rect class=\"sym\" height=\"6\" style=\"fill:purple;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a19 c2 node n18 p0\" transform=\"translate(-33.5417 20.1263)\"><g class=\"a18 i5 leaf node n10 p2 sample\" transform=\"translate(28.75 20.1049)\"><path class=\"edge\" d=\"M 0 0 V -20.1049 H -28.75\"/><rect class=\"sym\" height=\"6\" style=\"fill:purple;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">10</text></g><g class=\"a18 c2 node n15 p2\" transform=\"translate(-28.75 16.4114)\"><g class=\"a15 i1 leaf node n2 p2 sample\" transform=\"translate(-19.1667 3.69354)\"><path class=\"edge\" d=\"M 0 0 V -3.69354 H 19.1667\"/><rect class=\"sym\" height=\"6\" style=\"fill:purple;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a15 i2 leaf node n5 p2 sample\" transform=\"translate(19.1667 3.69354)\"><path class=\"edge\" d=\"M 0 0 V -3.69354 H -19.1667\"/><rect class=\"sym\" height=\"6\" style=\"fill:purple;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -16.4114 H 28.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:purple;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">15</text></g><path class=\"edge\" d=\"M 0 0 V -20.1263 H 33.5417\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">18</text></g><path class=\"edge\" d=\"M 0 0 V -21.7707 H -81.4583\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">19</text></g><path class=\"edge\" d=\"M 0 0 V -271.086 H 86.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">20</text></g><path class=\"edge\" d=\"M 0 0 V -8.31238 H 81.4583\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">21</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab\" transform=\"translate(0 -11)\">22</text></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colour_map = {0:\"red\", 1:\"blue\", 2:\"purple\"}\n",
    "node_colours = {u.id: colour_map[u.population] for u in ts.nodes()}\n",
    "for tree in ts.trees():\n",
    "    print(\"Tree on interval:\", tree.interval)\n",
    "    # The code below will only work in a Jupyter notebook with SVG output enabled.\n",
    "    display(SVG(tree.draw(node_colours=node_colours, width=500, height=400)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Admixtures and population splits are special types of demographic events that affect the state of some of the defined populations, in addition to moving lineages between populations. The output below shows that by adding the admixture event, we are triggering a change in the state of `AdmixedPop` at `time = 50`; the population is active at the start of the simulation, but becomes inactive for all steps of the simulation beyond time 50."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (3)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>AncestralPop0</td><td>Plotted in red.</td><td>1000.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>AncestralPop1</td><td>Plotted in blue.</td><td>1000.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>AdmixedPop</td><td>Plotted in purple.</td><td>700.0</td><td>0</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Migration matrix</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>AncestralPop0</th><th>AncestralPop1</th><th>AdmixedPop</th></tr></thead><tbody><tr><td>AncestralPop0</td><td>0</td><td title='Backwards in time migration rate from population AncestralPop0 to AncestralPop1\n",
       "= 0.001 per generation. Forwards in time, this is the expected number of\n",
       "migrants moving from AncestralPop1 to AncestralPop0 per generation, divided by\n",
       "the size of AncestralPop0.'>0.001</td><td title='Backwards in time migration rate from population AncestralPop0 to AdmixedPop =\n",
       "0.0 per generation. Forwards in time, this is the expected number of migrants\n",
       "moving from AdmixedPop to AncestralPop0 per generation, divided by the size of\n",
       "AncestralPop0.'>0</td></tr><tr><td>AncestralPop1</td><td title='Backwards in time migration rate from population AncestralPop1 to AncestralPop0\n",
       "= 0.001 per generation. Forwards in time, this is the expected number of\n",
       "migrants moving from AncestralPop0 to AncestralPop1 per generation, divided by\n",
       "the size of AncestralPop1.'>0.001</td><td>0</td><td title='Backwards in time migration rate from population AncestralPop1 to AdmixedPop =\n",
       "0.0 per generation. Forwards in time, this is the expected number of migrants\n",
       "moving from AdmixedPop to AncestralPop1 per generation, divided by the size of\n",
       "AncestralPop1.'>0</td></tr><tr><td>AdmixedPop</td><td title='Backwards in time migration rate from population AdmixedPop to AncestralPop0 =\n",
       "0.0 per generation. Forwards in time, this is the expected number of migrants\n",
       "moving from AncestralPop0 to AdmixedPop per generation, divided by the size of\n",
       "AdmixedPop.'>0</td><td title='Backwards in time migration rate from population AdmixedPop to AncestralPop1 =\n",
       "0.0 per generation. Forwards in time, this is the expected number of migrants\n",
       "moving from AncestralPop1 to AdmixedPop per generation, divided by the size of\n",
       "AdmixedPop.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events (1)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>200</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_admixture'>Admixture</a></td><td>derived=AdmixedPop ancestral=[AncestralPop0, AncestralPop1] proportions=[0.60, 0.40]</td><td>Moves all lineages from admixed population &#x27;AdmixedPop&#x27; to ancestral populations. Lineages move to &#x27;AncestralPop0&#x27; with proba 0.6; &#x27;AncestralPop1&#x27; with proba 0.4. Set &#x27;AdmixedPop&#x27; to inactive, and all migration rates to and from &#x27;AdmixedPop&#x27; to zero.</td></tr></tbody></table></div></div>"
      ],
      "text/plain": [
       "Demography(populations=[Population(initial_size=1000, growth_rate=0, name='AncestralPop0', description='Plotted in red.', extra_metadata={}, default_sampling_time=None, initially_active=None, id=0), Population(initial_size=1000, growth_rate=0, name='AncestralPop1', description='Plotted in blue.', extra_metadata={}, default_sampling_time=None, initially_active=None, id=1), Population(initial_size=700, growth_rate=0, name='AdmixedPop', description='Plotted in purple.', extra_metadata={}, default_sampling_time=None, initially_active=None, id=2)], events=[Admixture(time=200, derived='AdmixedPop', ancestral=['AncestralPop0', 'AncestralPop1'], proportions=[0.6, 0.4])], migration_matrix=array([[0.   , 0.001, 0.   ],\n",
       "       [0.001, 0.   , 0.   ],\n",
       "       [0.   , 0.   , 0.   ]]))"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dem"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This means that, for example, adding any demographic events that affect the AdmixedPop before this time will produce an error:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # THIS WON'T WORK\n",
    "# dem.add_migration_rate_change(time=80, rate=0.01, source=\"AncestralPop0\", dest=\"AdmixedPop\");\n",
    "# ts = msprime.sim_ancestry(\n",
    "#   samples={\"AncestralPop0\" : 0, \"AncestralPop1\" : 0, \"AdmixedPop\" : 6},\n",
    "#   demography=dem,\n",
    "#   sequence_length=1000,\n",
    "#   random_seed=63,\n",
    "#   recombination_rate=1e-7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Population splits\n",
    "\n",
    "(11.25am)\n",
    "\n",
    "We can also simulate population divergences with msprime. Suppose we want to model a situation where all lineages from multiple populations are migrants from a single ancestral population at a single point in time.\n",
    "\n",
    "<img src=\"pics/worksheet2-dem4.png\" width=\"400\" height=\"400\">\n",
    "\n",
    "Weâ€™ll specify this with the `add_population_split()` method. We need to know the time of the event, and the IDs of the derived and ancestral populations participating in the divergence event.\n",
    "\n",
    "Notice that in this case, we do not need to provide proportions as we did in the case of admixture. This makes sense when you consider that `msprime` simulates backwards-in-time: all lineages in all of the derived populations originate from the ancestral population in a split event. Any differences in â€˜quantitiesâ€™ of migrants must be modelled by sizes of the derived populations at the time of the split."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (3)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>R</td><td>Plotted in red.</td><td>500.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>B</td><td>Plotted in blue.</td><td>500.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>AncestralPopulation</td><td>Plotted in green.</td><td>500.0</td><td>0</td><td>1e+02</td><td>{}</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Migration matrix (all zero)</b><table border=\"1\" class=\"tskit-table\"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events (1)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>100</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[R, B], ancestral=AncestralPopulation</td><td>Moves all lineages from derived populations &#x27;R&#x27; and &#x27;B&#x27; to the ancestral &#x27;AncestralPopulation&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div>"
      ],
      "text/plain": [
       "Demography(populations=[Population(initial_size=500, growth_rate=0, name='R', description='Plotted in red.', extra_metadata={}, default_sampling_time=None, initially_active=None, id=0), Population(initial_size=500, growth_rate=0, name='B', description='Plotted in blue.', extra_metadata={}, default_sampling_time=None, initially_active=None, id=1), Population(initial_size=500, growth_rate=0, name='AncestralPopulation', description='Plotted in green.', extra_metadata={}, default_sampling_time=100, initially_active=False, id=2)], events=[PopulationSplit(time=100, derived=['R', 'B'], ancestral='AncestralPopulation')], migration_matrix=array([[0., 0., 0.],\n",
       "       [0., 0., 0.],\n",
       "       [0., 0., 0.]]))"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dem = msprime.Demography()\n",
    "dem.add_population(name=\"R\", description=\"Plotted in red.\", initial_size=500)\n",
    "dem.add_population(name=\"B\", description=\"Plotted in blue.\",initial_size=500)\n",
    "dem.add_population(name=\"AncestralPopulation\", description=\"Plotted in green.\", initial_size=500)\n",
    "\n",
    "# Add the population split.\n",
    "dem.add_population_split(time=100, derived=[\"R\",\"B\"], ancestral=\"AncestralPopulation\")\n",
    "dem"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Population splits will also modify the state of each of the derived populations, changing them from active to inactive at the time of the split."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populations of nodes from time < 100:\n",
      "[0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0]\n",
      "Populations of nodes from time >= 100:\n",
      "[2, 2, 2, 2, 2, 2, 2, 2]\n"
     ]
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "  samples={\"R\" : 3, \"B\" : 3, \"AncestralPopulation\" : 0},\n",
    "  demography=dem,\n",
    "  sequence_length=1000,\n",
    "  random_seed=63\n",
    ")\n",
    "\n",
    "print(\"Populations of nodes from time < 100:\")\n",
    "print([u.population for u in ts.nodes() if u.time < 100])\n",
    "print(\"Populations of nodes from time >= 100:\")\n",
    "print([u.population for u in ts.nodes() if u.time >= 100])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Changing population sizes or growth rates\n",
    "\n",
    "(11.35am))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also may wish to specify changes to rates of population growth, or sudden changes in population size at a particular time. Both of these can be specified by applying the `add_population_parameters_change()` method to our `Demography` object.\n",
    "\n",
    "<img src=\"pics/worksheet2-dem5.png\" width=\"400\" height=\"400\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "dem = msprime.Demography()\n",
    "dem.add_population(name=\"R\", description=\"Plotted in red.\", initial_size=500)\n",
    "dem.add_population(name=\"B\", description=\"Plotted in blue.\",initial_size=500)\n",
    "dem.set_migration_rate(source=0, dest=1, rate=0.05)\n",
    "dem.set_migration_rate(source=1, dest=0, rate=0.02)\n",
    "\n",
    "# Bottleneck in Population 0 between 50 - 150 generations ago.\n",
    "dem.add_population_parameters_change(time=50, initial_size=250, population=0)\n",
    "dem.add_population_parameters_change(time=150, initial_size=500, population=0)\n",
    "\n",
    "# Exponential growth in Population 1 starting 50 generations ago.\n",
    "dem.add_population_parameters_change(time=100, growth_rate=0.01, population=1)\n",
    "\n",
    "# Sort events, since we've added some out of time order.\n",
    "dem.sort_events()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Simulate.\n",
    "ts = msprime.sim_ancestry(\n",
    "    samples={\"R\" : 3, \"B\" : 3},\n",
    "    demography=dem,\n",
    "    sequence_length=1000,\n",
    "    random_seed=63461\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that because `msprime` simulates backwards-in-time, parameter changes must be interpreted backwards-in-time as well. For instance, the population growth event in the example above specifies continual growth in the early history of population 1 up until 100 generations in the past."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Debugging demography\n",
    "\n",
    "(11.40am)\n",
    "\n",
    "To help you spot any mistakes in your specified demography, `msprime` provides a debugger that prints out your population history in a more human-readable form. Itâ€™s good to get into the habit of running this debugger before running your simulations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><details open=\"true\"><summary>Epoch[0]: [0, 50) generations</summary><div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (total=2 active=2)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>R</th><th>B</th></tr></thead><tbody><tr><td>R</td><td> 500.0</td><td> 500.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population R to B = 0.05 per generation.\n",
       "Forwards in time, this is the expected number of migrants moving from B to R per\n",
       "generation, divided by the size of R.'>0.05</td></tr><tr><td>B</td><td> 500.0</td><td> 500.0</td><td> 0</td><td title='Backwards in time migration rate from population B to R = 0.02 per generation.\n",
       "Forwards in time, this is the expected number of migrants moving from R to B per\n",
       "generation, divided by the size of B.'>0.02</td><td>0</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events @ generation 50</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>50</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_parameters_change'>Population parameter change</a></td><td>population=0, initial_size=250</td><td>initial_size â†’ 2.5e+02 for population 0</td></tr></tbody></table></div></div></details><details open=\"true\"><summary>Epoch[1]: [50, 100) generations</summary><div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (total=2 active=2)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>R</th><th>B</th></tr></thead><tbody><tr><td>R</td><td> 250.0</td><td> 250.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population R to B = 0.05 per generation.\n",
       "Forwards in time, this is the expected number of migrants moving from B to R per\n",
       "generation, divided by the size of R.'>0.05</td></tr><tr><td>B</td><td> 500.0</td><td> 500.0</td><td> 0</td><td title='Backwards in time migration rate from population B to R = 0.02 per generation.\n",
       "Forwards in time, this is the expected number of migrants moving from R to B per\n",
       "generation, divided by the size of B.'>0.02</td><td>0</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events @ generation 100</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>100</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_parameters_change'>Population parameter change</a></td><td>population=1, growth_rate=0.01</td><td>growth_rate â†’ 0.01 for population 1</td></tr></tbody></table></div></div></details><details open=\"true\"><summary>Epoch[2]: [100, 150) generations</summary><div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (total=2 active=2)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>R</th><th>B</th></tr></thead><tbody><tr><td>R</td><td> 250.0</td><td> 250.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population R to B = 0.05 per generation.\n",
       "Forwards in time, this is the expected number of migrants moving from B to R per\n",
       "generation, divided by the size of R.'>0.05</td></tr><tr><td>B</td><td> 500.0</td><td> 303.3</td><td> 0.01</td><td title='Backwards in time migration rate from population B to R = 0.02 per generation.\n",
       "Forwards in time, this is the expected number of migrants moving from R to B per\n",
       "generation, divided by the size of B.'>0.02</td><td>0</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events @ generation 150</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>150</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_parameters_change'>Population parameter change</a></td><td>population=0, initial_size=500</td><td>initial_size â†’ 5e+02 for population 0</td></tr></tbody></table></div></div></details><details open=\"true\"><summary>Epoch[3]: [150, inf) generations</summary><div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (total=2 active=2)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>R</th><th>B</th></tr></thead><tbody><tr><td>R</td><td> 500.0</td><td> 500.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population R to B = 0.05 per generation.\n",
       "Forwards in time, this is the expected number of migrants moving from B to R per\n",
       "generation, divided by the size of R.'>0.05</td></tr><tr><td>B</td><td> 303.3</td><td> 0.0</td><td> 0.01</td><td title='Backwards in time migration rate from population B to R = 0.02 per generation.\n",
       "Forwards in time, this is the expected number of migrants moving from R to B per\n",
       "generation, divided by the size of B.'>0.02</td><td>0</td></tr></tbody></table></div></div></details></div>"
      ],
      "text/plain": [
       "<msprime.demography.DemographyDebugger at 0x13860c860>"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_history = msprime.DemographyDebugger(demography=dem)\n",
    "my_history"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can also use the debugger to perform some informative calculations based on the given demography."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/miniconda3/lib/python3.7/site-packages/msprime/demography.py:4321: UserWarning: Doubling the number of steps has resulted in different  predictions, please re-run with smaller step sizes to ensure  numerical accuracy.\n",
      "  \"Doubling the number of steps has resulted in different \"\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZQAAAEGCAYAAABCa2PoAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAABCUklEQVR4nO3dd3gV1dbA4d9KJ5QQQkAgQAIEpYQiEbCABVGwgIIIXhFQFK8F27Xg1c+C6LV3vYqCghcFRFFsYENERSmCNCmhSUILEEIJpK7vj5lgCAk5JOfkJGS9z3Me5uzZs2edSXIWM3tmb1FVjDHGmLIK8HcAxhhjTgyWUIwxxniFJRRjjDFeYQnFGGOMV1hCMcYY4xVB/g7An+rWrauxsbH+DsMYYyqVRYsW7VTV6MLlVTqhxMbGsnDhQn+HYYwxlYqIbCqq3C55GWOM8QpLKMYYY7zCEooxxhivqNJ9KEXJzs4mOTmZQ4cO+TsU44GwsDBiYmIIDg72dyjGVHmWUApJTk6mZs2axMbGIiL+Dsccg6qya9cukpOTiYuL83c4xlR5dsmrkEOHDhEVFWXJpBIQEaKiouxs0pgKwhJKESyZVB72szKm4rBLXsYYU0Wkpafyyow7yMzJoG/iLXROON+r7dsZSgWzZ88eXn/9da+3GxsbS0JCAu3atePss89m06Yin0sqV8OGDWPatGn+DsOYKiEnJ5u7Jl/Ch3lLmRGQxJrkRV7fhyWUCsZXCQVg9uzZLF26lHPOOYcxY8b4ZB/GmIrpgYn9WBiWwYCAdiy++ncG9bzL6/uwhFLBjBo1inXr1tGhQwfuuecebrnlFmbMmAHA5ZdfznXXXQfA+PHjeeCBBwB4/vnnadu2LW3btuXFF18scR+nn346KSkpR5Xn5uYybNgw2rZtS0JCAi+88AIAb731Fqeddhrt27enf//+ZGRkAM4Zxk033UTXrl1p1qwZP/zwA9dddx2tWrVi2LBhh9utUaMGd955J23atKFHjx6kpqYete9FixZx9tln06lTJy688EK2bt0KwMsvv0zr1q1p164dgwYN8vxAGmMOe+XDO/kycCPdMuvw4D8mEhQUTFCQ92+192kfioj0Al4CAoG3VfXJQutDgYlAJ2AXMFBVN7rr7geGA7nAbao6yy0fD1wC7FDVtkXs81/As0C0qu4sS/yPfraClVv2lqWJo7RuWIuHL21T7Ponn3yS5cuXs2TJEgAmT57M3Llz6dOnDykpKYe/aOfOncugQYNYtGgR77zzDr/99huqSpcuXTj77LPp2LFjsfuYOXMml1122VHlS5YsISUlheXLlwPO2RJAv379uOGGGwB48MEHGTduHCNHjgQgLS2NefPmMWPGDPr06cPPP//M22+/zWmnncaSJUvo0KEDBw4cIDExkRdeeIHRo0fz6KOP8uqrrx7eb3Z2NiNHjuTTTz8lOjqaKVOm8MADDzB+/HiefPJJNmzYQGho6OF4jDGe+/SHsby7/xtaZQfxzODPCQgM9Nm+fHaGIiKBwGtAb6A1cJWItC5UbTiQpqotgBeAp9xtWwODgDZAL+B1tz2Ad92yovbZGLgA+MurH8aPunXrxty5c1m5ciWtW7emfv36bN26lXnz5nHGGWfw008/cfnll1O9enVq1KhBv379mDt3bpFtnXvuuTRq1IivvvqKq6666qj1zZo1Y/369YwcOZKZM2dSq1YtAJYvX063bt1ISEhg0qRJrFix4vA2l156KSJCQkIC9evXJyEhgYCAANq0acPGjRsBCAgIYODAgQAMHjyYn3766Yj9rl69muXLl9OzZ086dOjAmDFjSE5OBqBdu3ZcffXV/O9//yMoyO4hMeZ4zF/2Lc+sf4m6ufDUxZOpHl7Tp/vz5V9oZyBJVdcDiMhkoC+wskCdvsAj7vI04FVx7gPtC0xW1Uxgg4gkue3NU9UfRSS2mH2+ANwLfOqND3CsM4ny0qhRI/bs2cPMmTPp3r07u3fvZurUqdSoUYOaNY/vl2P27NnUrl2bq6++mocffpjnn3/+iPWRkZH88ccfzJo1izfeeIOpU6cyfvx4hg0bxieffEL79u159913+eGHHw5vExoaCjhJI385/31OTk6RcRS+1VdVadOmDfPmzTuq7hdffMGPP/7IZ599xuOPP86yZcsssRjjgU1b1vB/v90BAfBol+eJa3SKz/fpyz6URsDmAu+T3bIi66hqDpAORHm47RFEpC+Qoqp/lFBvhIgsFJGFRV3L97eaNWuyb9++I8q6du3Kiy++SPfu3enWrRvPPvss3bp1A5wzmE8++YSMjAwOHDjA9OnTD68rSlBQEC+++CITJ05k9+7dR6zbuXMneXl59O/fnzFjxvD7778DsG/fPho0aEB2djaTJk067s+Ul5d3+G6u999/n7POOuuI9SeffDKpqamHE0p2djYrVqwgLy+PzZs3c+655/LUU0+Rnp7O/v37j3v/xlQ1+w7s4e7PryQ1CP4VN5KuCReUy35PiE55EQkH/g08VFJdVR2rqomqmhgdfdT8MH4XFRXFmWeeSdu2bbnnnnsAJ2nk5OTQokULTj31VHbv3n04aZx66qkMGzaMzp0706VLF66//vpj9p8ANGjQgKuuuorXXnvtiPKUlBTOOeccOnTowODBg/nPf/4DwGOPPUaXLl0488wzOeWU4/9fTvXq1Zk/fz5t27bl+++/56GHjvwxhYSEMG3aNO677z7at29Phw4d+OWXX8jNzWXw4MEkJCTQsWNHbrvtNmrXrn3c+zemKsnJyebO//VmVWguw2v15vJz/1lu+xZV9U3DIqcDj6jqhe77+wFU9T8F6sxy68wTkSBgGxANjCpYt2A9930s8Hl+p7yIJADfARlu0zHAFqCzqm4rLsbExEQtPMHWn3/+SatWrcr24c0RatSo4dMzC/uZGfO3+8ZfypeBG7mcUxg99EOf7ENEFqlqYuFyX56hLADiRSROREJwOtlnFKozAxjqLl8BfK9OhpsBDBKRUBGJA+KB+cXtSFWXqWo9VY1V1VicS2SnHiuZGGPMieb5KTe7twdH8sjgyeW+f58lFLdP5FZgFvAnMFVVV4jIaBHp41YbB0S5ne538feZyQpgKk4H/kzgFlXNBRCRD4B5wMkikiwiw331GYx3WL+HMb73/qzneO/gjyQcCubZIV/59Pbg4vj0dhlV/RL4slDZQwWWDwEDitn2ceDxIsqPvt/16DqxxxurMcZUVj8snM7LW8bTKCeAZ/tNJzysul/isPsvjTGmElu65hce/eNBwoAnzn6ThtFN/RaLJRRjjKmkNm1Zw30/juBQADyRMJp2Lc/wazwnxG3DxhhT1ezas427Ph/AjkD4V9ObOfe0/v4OyRJKRePr4es7dOhAQkICn37qlcEEysSGrzemdDIOHeCOKZeQFJLLjXX6ckWPW/wdEmAJpcLx9fD1S5YsYdq0adx2220+2YcxxrfycnO5a+KFLAnL5OqQ0xnR96h7l/zGEkoFUx7D1+/du5fIyMijym34emMqvvve7cPPoelcmteCe//xlr/DOYJ1yh/LV6Ng2zLvtnlSAvR+stjVvhy+/txzz0VVWb9+PVOnTj1qvQ1fb0zFNua9IcwM+ouzs6IYc13Fu1xsZygVnDeHr589ezbLly9n2bJl3HrrrUc9cGjD1xtTcb0y7S6m5C0m8VA4zw+d5ZcHF0tif6HHcowzifLizeHr8zVv3pz69euzcuVKOnfufLjchq83pmJ6e8bDjNv/Na2ygnjx6q8ICQkteSM/sDOUCsbXw9cD7Nixgw0bNtC06ZEPQNnw9cZUPO/Peo7Xd39Es6wAXun/BRE16vg7pGLZf/UqmILD1/fu3ZtnnnmGbt268fXXX9OiRQuaNm1a7PD1wDGHrz/33HMJDAwkOzubJ598kvr16x+xPiUlhWuvvZa8vDyAo4avj46OpkuXLkclvJLkD18/ZswY6tWrx5QpU45Ynz98/W233UZ6ejo5OTnccccdtGzZksGDB5Oeno6q2vD1psr59IexvLDlHRrmCC9d+hH1o445LZTf+Wz4+srAhq8vHzZ8vTHH7+t5H/Dwn2OolSe80vN/tGzawd8hHeaP4euNMcaUwk+LP2f0n2MIV3im+9gKlUyOxS55GZ+zfg9jPPf7yjk8+Pt9BACPd37B7+NzHQ9LKMYYU0GsWLeQ+365hewAGJMwutzmgvcWu+RljDEVwOoNi7l79jD2Bir/bnlfhRjs8XhZQjHGGD9bs2kJd3x3DbsClfuajuTis4b4O6RSsYRijDF+tGbTUu74erCTTJrcQr/z/unvkErNpwlFRHqJyGoRSRKRUUWsDxWRKe7630QktsC6+93y1SJyYYHy8SKyQ0SWF2rrGRFZJSJLRWS6iNT25Wc7UWRmZnL++efToUOHo54P8ZYaNWr4pF1jKrukv5Zz59dXkxoE9zS+if49bvZ3SGXis4QiIoHAa0BvoDVwlYi0LlRtOJCmqi2AF4Cn3G1bA4OANkAv4HW3PYB33bLCvgHaqmo7YA1wv1c/0Alq8eLFgDMwZP54W8YY31u/eQV3zLqK7UHK3THXM+D8W/0dUpn58gylM5CkqutVNQuYDPQtVKcvMMFdngb0EGegp77AZFXNVNUNQJLbHqr6I7C78M5U9WtVzR886lcgxtsfqLwUNRz9xo0badWqFTfccANt2rThggsu4ODBgwCsW7eOXr160alTJ7p168aqVauOanP37t1cdtlltGvXjq5du7J06VJ27NjB4MGDWbBgAR06dGDdunVHbFNcu5999hldunShY8eOnH/++Wzfvh1wbg++9tprSUhIoF27dnz00UeH23rggQdo3749Xbt2PVzfmKpqQ8oqbp85iG1Byt2NhjOw5x3+DskrfHnbcCNgc4H3yUCX4uqoao6IpANRbvmvhbY9njEHrgOKvH4jIiOAEQBNmjQ5ZiNPzX+KVbuP/nIui1PqnMJ9ne8rdn1xw9FHRkaydu1aPvjgA9566y2uvPJKPvroIwYPHsyIESN44403iI+P57fffuPmm2/m+++/P6Ldhx9+mI4dO/LJJ5/w/fffM2TIEJYsWcLbb7/Ns88+y+eff35ULMW1e9ZZZ/Hrr78iIrz99ts8/fTTPPfcczz22GNERESwbJkz5H9aWhoABw4coGvXrjz++OPce++9vPXWWzz44INePKrGVB6btqzhti8HsDVIuavhtQy64E5/h+Q1J9xzKCLyAJADFDmKoaqOBcaCM/RKOYbmkYLD0QOHh6Pv06cPcXFxdOjQAYBOnTqxceNG9u/fzy+//MKAAQMOt5GZmVlku/lnDOeddx67du1i7969xcZxrHaTk5MZOHAgW7duJSsri7i4OAC+/fZbJk+efLh+/iReISEhXHLJJYfj/uabb477uBhzIli/eQW3zxzEliDlzgZD+MeF//J3SF7ly4SSAjQu8D7GLSuqTrKIBAERwC4Ptz2KiAwDLgF6qBcGKTvWmYQ/FBwePjAwkIMHD5KXl0ft2rUPT8jlLcdqd+TIkdx111306dOHH374gUceeeSYbQUHBx8esj4wMLDYYe2NOZGt3rCYO7+9hh1BcFeDIVzd615/h+R1vuxDWQDEi0iciITgdLLPKFRnBjDUXb4C+N5NBDOAQe5dYHFAPDD/WDsTkV7AvUAfVc3w4ucoV8c7HH2tWrWIi4vjww8/BJy5Rf74448i280fev6HH36gbt26hyfQOt5209PTadTIuQI5YcKEw9v07NmT11577fD7/EtexlR1S9f8wm3fXcOuIOW+xv88IZMJ+DChuB3ktwKzgD+Bqaq6QkRGi0gft9o4IEpEkoC7gFHutiuAqcBKYCZwi6rmAojIB8A84GQRSRaR4W5brwI1gW9EZImIvOGrz+ZLBYej79KlyzGHo883adIkxo0bR/v27WnTpg2ffvrpUXUeeeQRFi1aRLt27Rg1atQRieB4233kkUcYMGAAnTp1om7duofrP/jgg6SlpdG2bVvat2/P7Nmzj/PTG3PiWbRiNnf9eAPpgcq/m915QtzNVRwbvt6Gr6/07GdmKqpf/viKfy+8m2yB/zvl3/Q642p/h+QVxQ1ff8J1yhtjTEUwe8FHPLzsIRBnoMfKODbX8bKEYowxXjbrl0mMXvUEwQpjOj3DWR0u8ndI5cISShFU9fBdSaZiq8qXbE3FNO27V3n2r/9SXeHJri9zWpse/g6p3NjgkIWEhYWxa9cu+6KqBFSVXbt2ERYW5u9QjAFg/GeP8uRfbxCZK7zQfVyVSiZgZyhHiYmJITk5mdTUVH+HYjwQFhZGTEylHWXHnEBenDqSdzNm0yw7gOcunkpco1P8HVK5s4RSSHBw8OEnv40xxhOPTfwHU3UZCZkhvHTl50RHNvR3SH5hCcUYY0opLzeXUe/24augvzjtUHVeGjyTmtVr+zssv7GEYowxpZCVdYg7372AH0PTODsriuevnUVISGjJG57ArFPeGGOO0559O7nxne78GJrGxblxvHzdd1U+mYAlFGOMOS6bUlZz/Qc9WBSawcCAjjx53QwCAgNL3rAKsIRijDEeWrjiB278qj8bgnO5JeJSHrxmor9DqlCsD8UYYzzw5U8TeXLNU+QGwIOxI7n83H/6O6QKxxKKMcaU4J3PH+P11ClE5MHo057ljPa9/R1ShWQJxRhjjuHp929gUtY84rIDeOrCiZzctIO/Q6qwLKEYY0wRcnKy+feEy/gq6C/aZ4bxwpWfER3ZwN9hVWiWUIwxppC09FT+NeUSFoRm0D0ziueGfUlYaLi/w6rw7C4vY4wp4M/1i7h2ag8WhhxggLTjleHfWTLxkJ2hGGOMa9a89/nPysc5GAR3RA3kuksf8ndIlYolFGOMAcZ++gBv7f6EWsB/2jzKeZ2v8HdIlY5Hl7xEpKmInO8uVxORmh5u10tEVotIkoiMKmJ9qIhMcdf/JiKxBdbd75avFpELC5SPF5EdIrK8UFt1ROQbEVnr/hvpSYzGmKotLzeXhyYM4JU9M2iSHcTr50+yZFJKJSYUEbkBmAa86RbFAJ94sF0g8BrQG2gNXCUirQtVGw6kqWoL4AXgKXfb1sAgoA3QC3jdbQ/gXbessFHAd6oaD3znvved3ByfNm+M8b30/bu5aVx3prOKLodqMv4fs+224DLw5AzlFuBMYC+Aqq4F6nmwXWcgSVXXq2oWMBnoW6hOX2CCuzwN6CHO3Lt9gcmqmqmqG4Aktz1U9UdgdxH7K9jWBOAyD2IslXljR7Lh8VN91bwxphwsT/qNYe+fwy+he+mTF8/Y6+cSUaOOv8Oq1DxJKJluQgBARIIAT+bHbQRsLvA+2S0rso6q5gDpQJSH2xZWX1W3usvbgPpFVRKRESKyUEQWlnZWxoOh9YjL20T2jrWl2t4YU3YHDx3gvvGXcsHbbVn31/KSNyhg+uw3uPnH69gSnMttkZfz+LUf2wCPXuBJQpkjIv8GqolIT+BD4DPfhlU26kwIX2TSU9WxqpqoqonR0dGlav9g7PkA7F/2ealjNMaU3u9//sjVE8/gy8CNbA0WFq7+1qPt8nJzefL94Yze9CrV8oRn2/+HG/qM9nG0VYcnCWUUkAosA24EvlTVBzzYLgVoXOB9jFtWZB33zCcC2OXhtoVtF5EGblsNgB0exFgqtRvFszovhoC1s3y1C2NMMV77+B5u/vUmtgTncGlePAA70v8qcbt9B/Zw67hzmZQ9nzaZYYzr8zndTu3j63CrFE8SykhVfUtVB6jqFar6lojc7sF2C4B4EYkTkRCcTvYZherMAIa6y1cA37tnFzOAQe5dYHFAPDC/hP0VbGso8KkHMZZKTGQ1vss7lZrbF8ChdF/txhhTwM49W7hpbDfe2DeThtlBvN71v9x8yfMApGVsO+a2f65fxLBJZzM3NI2LcmMZf90vxNSLLYeoqxZPEsrQIsqGlbSR2ydyKzAL+BOYqqorRGS0iOT/t2AcECUiScBduHdmqeoKYCqwEpgJ3KKquQAi8gEwDzhZRJJFZLjb1pNATxFZC5zvvveJBhHV+D6vIwGaA0nf+Wo3xhjXrHnvM3jaBfwc4iSE/w39hVNbdadhncaE5inpWUXdp+P4ePbr/HP2EDYH53BLRB+euu4zQoJDyjH6qqPYBxtF5CrgH0CciBQ8s6hJ0XdZHUVVvwS+LFT2UIHlQ8CAYrZ9HHi8iPKriqm/C+jhSVxlFRIUwNYabTmQG0H1NbOgbb/y2K0xVU5ubjZPvH8tH+cuobYoDzQawcCef18gCQgMJDIX9ubtPWrbvNxcHps0mOl5y6ivwiMJozk30f5WfelYT8r/AmwF6gLPFSjfByz1ZVCVQYPIGizam0j3tbMgNxsCg/0dkjEnlOVJv/HEdzexLCybUzPDefTS/xHbsOVR9SLyAtnHwSPKknds5MHpV7Io7CCJmdV5ov80GtRtfNS2xruKTSiqugnYBJxefuFUHjGR1fgirTPds76D9XMg/nx/h2TMCSEvN5fXpt/N+3u/JidEuDr4dO4d/Eaxt/XW1DC2Bxw4/P6bXz/gmeWPkxoKVwZ05IHr37VbgstJiWN5iUhX4BWgFRACBAIHVLWWj2Or0GIiwxm3vxVP1qyFLP/IEooxXrAhZRWPfTGUBaEZnJwTxL1nPE/ntse+kl0roAarA/ejeXk8P/UW3j84l1qiPBx7O5edM6KcIjfg2eCQr+LcofUhkAgMAY4+76xiYiKrcTAviIzmF1F91eeQ/QIEh/k7LGMqrXc/H8347VPYFyJcThseHDaRkJDQErerFVyHfbKDG9/uxrzQvbTNDmFM7/do3rhNOURtCvJocEhVTQICVTVXVd+h6LG0qpRGkdUA+Kthb8jcC0mePVhljDnS9l3J3DK2O8/t+pCIvACebf0Io4dO8SiZAERWc0aC+jUknUtymzFh2DxLJn7iyRlKhvscyRIReRqno77KT8wVE+lMuLMytAOtwqNg+UfQ6hI/R2VM5TLt21d4Y+MbpIYIvXJieejqSdQMjziuNrq3HcAvP/9Mn5hBDLnIt2PCmmPzJKFcg5NAbgXuxHmCvb8vg6oMGtZ2Lm8lp2dD68tgyfvOQ45hx/fHYExVtH3XZsZ8PIQfQnbSAOGR2Nu4vJT9HYltzmFamyXeDdCUyjETijtk/BOqejVwCHi0XKKqBEKDAqlfK5TktAw442pYOA6WfQinXe/v0Iyp0CZ++QTvbp3ErmChZ3YjHhzwHnUiPBnA3FR0x0woqprrTq4VUnDEYeNoVLsayWkHoWEXOCkBFr0LicNBxN+hGVPhbN66hjGfXcsvoXtpnBfAY/F30af7tf4Oy3iRJ5e81gM/u0/LH77ZW1Wf91lUlURMZDiLN6c5CeTUofDl3bBlMTSyuVKMOUyVsZ/+m//tmsG+EOGi3OY8+I8J1Kxe29+RGS/zpHN9HfC5W7dmgVeV17hONbbsOUR2bh60uxKCqsHvE0re0JgqYs3GJVw3tguvpH9OZF4gT7d6mKeu+9SSyQmqxDMUVbV+k2LE1a1Bbp6yeXcGzaIjnDG9lk2DnqOtc95UaTnZWTz/4U1MP/QrWe5zJfcPeYdqoeH+Ds34UJW//bcs4upWB2DDTvdKYOcRkLXf6Usxpor67rcpDHwnkfey59MkJ4SXT32G0UOnWDKpAjzpQzHFaFY4oTTsAHHd4dc3oMtNEGRDZJuqI3V3Ck9+fB3fBaVQM0i5PvwCRg5+1sbRqkIsoZRBZPUQaocH/51QAM64HSb1dx507FDkSPvGnFhUGffZw0xK/YjU4AC6ZdXl/kvfpvFJLfwdmSlnJV7yEpGWIvKdiCx337cTkQd9H1rlEFe3+pEJpUUPqNcafnkFtMhp7Y05YSxd8zNDx3bmxbTphOcF8FjT23n9hh8smVRRnvShvAXcD2QDqOpSnMEiDU5CWZ9aIKGIwJm3w44V8GfhGY+NOTGk79vFg+/047qfb2RlyEH6SVumDvmVy86xB3urMk8SSriqFp7PPccXwVRGzaNrsG3vIfZnFjgkCQOg7snw/RjIy/VfcMZ4WV5uLm9/+n9cMaU7nwasJSG7BmPPeINHh3xAeGh1f4dn/MyThLJTRJoDCiAiV+AMEFkiEeklIqtFJElEjhq1TURCRWSKu/43EYktsO5+t3y1iFxYUpsi0kNEfheRJSLyk4iUyzl3fL0aACTt2P93YUAgnPcA7FwDS6eURxjG+NxPiz/nmnGn8dKeTwjRAB5ocD3vjPiVjief5e/QTAXhSaf8LcBY4BQRSQE2AINL2sgdB+w1oCeQDCwQkRmqurJAteFAmqq2EJFBwFPAQBFpjXNZrQ3QEPhWRPLnYCmuzf8CfVX1TxG5GXgQGObB5yuT+PrOM55rt++jQ+Paf69o1QcadIDZ/4E2/WyuFFNpbU3dxLOf3sD3QVsIC1IGBp7GXUNeszMScxRPHmxcD5wvItWBAFXd52HbnYEkd3tEZDLQFyiYUPoCj7jL04BXRUTc8smqmglsEJEktz2O0aYC+bNIRgBbPIyzTJrUCSckKIC1Bc9QnOCg56Mwsa/TQX/2PeURjjFek52VyavT72L6/h/YEySclRXFXb1fp4XNNWKK4ckUwE8AT6vqHvd9JPAvVS3pTq9GwOYC75OBLsXVUdUcEUkHotzyXwtt28hdLq7N64EvReQgsBfoWsznGQGMAGjSpEkJH6FkgQFC8+garN1eRJ5tdo5zpjL3OWg/EGqXfX/GlIcpXz/HpE3vsiEE4nODuO+Uu7n4zGv8HZap4DzpQ+mdn0wAVDUNuMhnEZXencBFqhoDvAMUOXilqo5V1URVTYyOjvbKjuPr1WDN9v1Fr7zwCedsZeb9XtmXMb700+LPGPJmImO2vkuGKCNq9GLqdYssmRiPeNKHEigioe7lJ0SkGuDJ3JwpOJNx5Ytxy4qqkywiQTiXqnaVsO1R5SISDbRX1d/c8inATA9i9IqTT6rJjD+2sPdQNrXCgo9cWbsxdL8bvhsNK2dA6z7lFZYxHlv31zJemnUbPwanEhaiXE4C/xr0OhE16vg7NFOJeJJQJgHficg77vtrAU+G1F0AxItIHE4yGAT8o1CdGcBQYB5wBfC9qqo7VP77IvI8Tqd8PDAfkGLaTAMiRKSlqq7B6bT/04MYvaJ1A6frZtXWfXSOK+IP8IzbYOWn8Pkd0OR0qOGdMyNjymrPvp288PFNzMxbSWaw0C27Pnf0eoXmjVv7OzRTCXnSKf+UiCwFerhFj6nqLA+2yxGRW4FZQCAwXlVXiMhoYKGqzgDGAe+5ne67cR+YdOtNxelszwFuUdVcgKLadMtvAD4SkTycBHOdx0ehjFo3dBLKyi3pRSeUwGC4/E1482wnqQz8n03CZfwqM+sgr0//F5/vn8OOoAA6ZoUzostozurQ29+hmUpMtAoPD5KYmKgLFy4sczuqSuKYb+nRqh5PX9G++Io/vwzf/B/0fga6lG7+bGPKIi83h3GfPcT0nTPYHCw0zxSuiruegT1v83dophIRkUWqmli43JO7vPrhPB9SD+eSkwCqqrWOuWEVIiK0bliLlVv3Hrvi6bfCpp9h1r+dkYkbdz52fWO8RZUp3zzHlI0TWRuqNFbh1tqXc8MlD9towMZrPLnL62mgj6pGqGotVa1pyeRorRvUYs22/WTl5BVfKSDAufQV0QimDoW9Hg04YEyZfPXTu1z11qmM2TqB9MA8hoWdw8dD5nNj39GWTIxXedIpv11Vy62Du7JKiIkgKzeP1dv2kRBzjNkaq9V2+lDG94JJA+DaLyHM8rPxvjkLpjHh96dZEHaQiKA8BgQkcsfAl6hVo7a/QzMnKE8SykIRmQJ8AmTmF6rqx74KqjJqH1MbgCXJe46dUABOSoArJ8D7A2HqNfCPqRDkyZ3YxpRszoJpTFj8DAtCMwgPyeOSvJO5re8rNIhqVPLGxpSBJwmlFpABXFCgTAFLKAXERFYjqnoIS/7awzVdm5a8QYvz4dKX4dObYco1MPA9SyqmTOYsmMbExc8wPzSD8OA8eue24Jbez9O0QXN/h2aqCE9uG762PAKp7ESEDo1r80fyHs836ng15ByCL+6CKYPhyokQXM1nMZoT05yFHzFx0dPMD7NEYvzLk7u8WuKM5FtfVduKSDucTvoxPo+ukmnfuDbfr95BekY2EeHBJW8AcNpw55mUz+90BpK8ajKE29PJpmQ/LvyICfmJJMQSifE/Ty55vQXcA7wJzoyNIvI+YAmlkNNi66AKCzftpker+p5vmHgdVIuEj2+EcT1h0PsQfbLvAjWV2lc/jmPqqjdYGHrIEompUDxJKOGqOl+OfLLbZmwsQscmtQkOFOZvPM6EAtDmcqjZACZfDWPPhUtecEYoNgZnpsSp3z7LJ5s+YEVoLjWD8rgoL56bez1nicRUGJ4klFLP2FjVhAUHktAoggUbdpeugSZd4Z8/wUfDYfoI5yHIXv+BEJvIqKrKzspkwhcP80XqFySFQlRgHv2lIzf2fZoGUQ39HZ4xR/DZjI1VVee4KMb9tJ6MrBzCQzw5vIXUagBDZsAPTzjzqKybDRc/By0vKHnbiionC7YtBQQCgyAg2BnfLCCwwHJwoXVBVXq8s4yMdMZ+dh+z9v1EcrDQIEC5OqQ7N1/xH2pVL+G2dGP8xJczNlZJZzSP4o056/htw27OPble6RoJDIIeD0GLns5gku8PcC6J9XzMGQ6/kkme+wzT/3iTHAHFSRJ54pzy/v0StFB5ngQ4/yIgQl5Ry0CeyOF6KvltSYG2nPr5+8h/4a7PX1DUWc5f7yq4XHAcBJUjVxa3jRZIjMXWKXTMdgTmsDMogFgNYETNixhxyaOEhtht5aZi8+WMjVVS57g6hAQF8NPanaVPKPmang43zoVfXoI5z8CqLyBxOHT7V6UaAv+tDXP4ODICyXNG+hH3S1sAcb9J5XCpU5Y/aNzhMkC0wHZHbX9kWcFtAwq1G5CfOPTvOvmNHG5P/m63wOqjy+To9UW992TZSXlOSeOcUIY06M+QC+4mMMiGRzGVgyfXZHqr6r/z36hqmohcBFhCKUJYcCCdY+swd22qdxoMCoHu90D7q2DOUzB/LCx+D04dCl1uhEgPHqL0s/TsvdRWePTMrxERAkQIEAgUcd87UynnLweIuO//Xg4QDm/rbAcBAc7y3+tKaMfdTqrwpTRjfMmXMzZWWd1b1uWJL1eRsucgjWp76UHFiBjo84ozWdecp2H+m/Dbf6HVpdDpWog72xl8sgLazyFq5AVy3inHeeebMaZS8eQbKH/GxuEiMhz4Bs9mbKyy8m8Z/nbldu83Xjce+r8Fty+FM2+H9XPgvcvgxQRnmuFty6GCzXGzT7IJV/s/iDEnuhITiqo+hfMQYyv39ZiqPu3rwCqz5tE1aBZdnW98kVDyRTSC8x+Bf62GK96Beq3gpxfgjTPhpXbw1Sgn2eRk+S4GT6iSHpBHdcL9G4cxxuc86ZSvDnytqjNF5GTgZBEJVtVs34dXefVsXZ9xczewJyOL2uEhvttRcBi07ee89m2HNV/Bqi9h4XjnklhwuDORV9OzIPZMaNixfMcLy9jN7sAAmmK3uhpzovOkD+VHoJt7d9dMYCEwELjal4FVdpckNOTNOev5avk2rurcpHx2WrM+dBrmvLIOOM+wbJwLG3+C2e5IORII0adAg/Z/v+q3hjDffOFn7NnEwYAAwgNtfDJjTnSeJBRR1Qy3/+S/qvq0iCzxpHER6QW8BAQCb6vqk4XWhwITgU7ALmCgqm50190PDAdygdtUddax2hTn1p0xwAB3m/+q6suexOkLbRvVoll0dT5ZnFJ+CaWgkOrQ6hLnBZCxG/6aB1sWw9Y/IOlb+OP9v+tXj4aoeIhqDlEtoE4c1IpxLq1Vr1fqDv/tO5MAqBF2Ulk/kTGmgvMooYjI6ThnJMPdshJvjBeRQOA1oCeQDCwQkRmqurJAteFAmqq2EJFBOHPXDxSR1sAgoA3QEPjWHfWYY7Q5DGgMnKKqeSJSxodAykZEuKxDI57/Zg2bd2fQuI6f+xDC68ApFzuvfPu2OckldRXsXAu71sGamXCg0C3PAcFQq6Fzp1nNkyC8LlSvC+FR7r8F3ofWcm51dm3ZtR6AWjViyuNTGmP8yJOEcgdwPzBdVVeISDNgtgfbdQaS3CftEZHJQF+gYELpCzziLk8DXnXPNPoCk91blTeISJLbHsdo8ybgH6qaB6CqOzyI0aeu6BTDi9+uYfKCv7jnwlP8Hc7Rap7kvFpeeGT5wT2w5y/YmwLpyc5rbwqkpzhnOAd2QWZ68e0GhTmX0EJrcTAvA2oHEhXZzKcfxRjjf54MvTIHmCMi4e779cBtHrTdCNhc4H0y0KW4OqqaIyLpQJRb/muhbfPnLy2uzeY4ZzeXA6k4l8nWFg5KREYAIwCaNPHtpaiGtatx3in1mLIgmdt7tCQkqGI+J3KUarWdV4N2xdfJyYKMXZCxEw7sdJd3O4nm0F44lA6Ze9mavhHYRYNIP1z2M8aUK0/u8jodGAfUAJqISHvgRlW92dfBHadQ4JCqJopIP2A80K1wJVUdizPYJYmJiT5/YGNw16Z8++cCPl2SwoDEyjcOV2FZuVnsPrSbtENppB1KIz0rnf25+8kIyGB/aDb7JY8DgcL+kCAOZIfyZ04wZEGTCL9egTTGlANPLnm9CFwIzABQ1T9EpLsH26Xg9Gnki3HLiqqTLCJBQARO5/yxti2uPJm/57mfDrzjQYw+d3bLaE45qSZv/rie/qfGEBBQMYf9UFV2H9rNtgPb2HJgC1v3b2XrAeeVmpHqJJHMNA5kHyi2DUGoHlyd6sHVqRFcw/k3oD7bd7WkQa1a5fhpjDH+4NH46qq6udD4R7kebLYAiBeROJwv/UHAPwrVmQEMBeYBVwDfq6qKyAzgfRF5HqdTPh6YjzOWXnFtfgKcizO8/tnAGk8+m6+JCDed05zbJy/hs6Vb6NuhUckb+ZCqsj1jO2vS1rB+z3qS9iSxPn096/asIyMn44i61YKq0bB6Q+qF1yOmZgx1wuoQGRZJZFgkdUKd5bDAGuzZH8T2PcqWtFy27MkkOe0gKXsOsiotg0PZeYQEBRBRzcMpkY0xlZYnCWWziJwBqIgEA7cDf5a0kdsnciswC+eusPFup/5oYKGqzsC5lPae2+m+GydB4NabitPZngPcoqq5AEW16e7ySWCSiNwJ7Aeu9+wQ+N6l7Rryxpz1PPv1anq1PYnQchw9Nicvh+U7l7N4x2KWpi5laepSdhz8+36FutXq0rx2cy5rcRlNajWhQfUGNKzRkAbVG1ArpNbhgRSzcvJYs30fK7fsZdVf+1ifup91qelsTtt6xEgvkeHBxESG0yK6Bue0jKZRZDXaN65NYAU9MzPGeI9oCeM+iUhdnOc+zsc5Q/gauF1Vd/k+PN9KTEzUhQsXlsu+flyTypDx87mrZ0tu6xHv031tP7Cdn7f8zE8pP/Hrll/Zl+1MYdO4ZmPaRbcjoW4CraNa0yyiGRGhRz/QmJWTx4ot6SxNTmfFlnRWbNnLmu37yM51flfCggNoVtcZXqZ5dA2a16tBs7rViatbneqhpZhUzBhTqYjIIlVNLFzuyV1eO7Gn4suse8toLmnXgFdnJ3Fhm5M4+aSaXm0/PTOdbzZ9wxfrv2DR9kUoSr3wevSM7ckZDc/gtJNOo05Y0U+rp2dk8/tfaSzctJsFG9P4Y/MeMnOcqaSiqofQumEthp/VjLaNatGmYQRN64RX2L4gY4z/FHuGIiKvcPREcoepqie3Dldo5XmGApC6L5PeL/1IneohfHLLmaWbIriQNWlrmPTnJL5Y/wWZuZnE1orlomYX0aNJD+Jrxxc598eh7Fzmb9jN3LWpzF27k1XbnDOYoAChTaMIEptGktg0ko5NIqlfK9TmDzHGHKE0Zyjl901bRUTXDOWFgR0YOn4+t32wmDcGdyIosHTPpizZsYRXl7zKb1t/IywwjD7N+9A/vj+to1oflQDy8pRV2/YdTiDzN+4mKyePkMAAEmMj+VfPliTG1qFD49pUC7HZAY0xpVNiH8qJrLzPUPJNnLeRhz5dwcUJDXh+YPvj6qRft2cdL/3+ErM3zyYqLIprWl/DFS2vOKovZOf+TOauTWXO6lR+StrJzv3OMPYn16/JWfF16RZfly5xUZZAjDHHrdR9KCISDdwHtAbC8stV9TyvRliFDDk9lqycPMZ88Sdb0w/y7ID2NIuuccxtDuYc5L9L/svElRMJCwpjZMeRDG41mPBgZ4yw7Nw8Fv+1hzlrdjBnTSrLU/YCTh9It/i6nBUfTbf4utSvFXas3RhjTKl5chF/EjAFuBj4J85zI16aML3qur5bMxpEVGPUR0vp9eJchneL49ozYqlXxBf+0tSljJo7is37NtM/vj+3n3o71YMiWLElnUWbtjF/w27mrdvFvswcAgOETk0iufuClpzdsh5tGtayDnRjTLnw5LbhRaraSUSWqmo7t2yBqp5WLhH6kL8ueRW0Y98hnvxyFR8vTiFA4MwWdUlsWoeTT6pJZHgw322dytR1b1IruC7nRt1K1v7mJO3Yx7KUdA5lO3diNa5TjbNa1OXsltGc0aIutcLsIUJjjO+U+pIXkD8z41YRuRjYAthsSV5Sr2YYzw/swK3nteDj31P4cvlWfkpag5JNWINpBEf8QfbetuzZ2p8JecFEhm8lrm51rurchNNi65DYNLLIsxpjjClvnpyhXALMxRlD6xWgFvCo+6R7pVYRzlCKsmXvLkZ+P5I16cu4OGY4Fze+mojwYBpHhhNZ3YfTCRtjjAfK8mDj5+5iOs5YWcaH0jPTuWPOTWzYl8QzZz9Dr9he/g7JGGM8UuJDECISIyLTRSRVRHaIyEciYtPv+UB6Zjo3fH2Dc2vwuS9ZMjHGVCqePFX3Ds6owA1wRv79jAoyNPyJJCs3izt/uJOkPUm8fN7LdIs5aioXY4yp0DxJKNGq+o6q5rivd4FoH8dVpagqD//yMAu2LeCxMx/jzEZn+jskY4w5bp4klF0iMlhEAt3XYJxJsIyXTFw5kc/Xf87IjiO5uNnF/g7HGGNKxZOEch1wJbAN2IozEda1vgyqKvkj9Q9eXPQi5zc5nxsSbvB3OMYYU2qe3OW1CehTDrFUOXuz9nLvnHupX70+j575qI3qa4yp1Dy5y2uCiNQu8D5SRMb7NKoq4tkFz7I9YztPd3+aWiE257oxpnLz5JJXO1Xdk/9GVdOAjj6LqIqYv3U+05OmM7TNUNpFt/N3OMYYU2aeJJQAEYnMfyMidfBsyBZEpJeIrBaRJBEZVcT6UBGZ4q7/TURiC6y73y1fLSIXHkebL4vIfk/i85dDOYd4dN6jNK7ZmJva3+TvcIwxxis8SQzPAfNE5EP3/QDg8ZI2EpFA4DWgJ5AMLBCRGaq6skC14UCaqrYQkUHAU8BAEWkNDALa4Dz78q2ItHS3KbZNEUkEIqngJqyYwF/7/uKtC94iLMjG4TLGnBhKPENR1YlAP2C7++qnqu950HZnIElV16tqFjAZ6FuoTl9ggrs8DeghTs90X2Cyqmaq6gYgyW2v2DbdBPYMcK8HsfnN7kO7eWfFO5zX+Dy6Nujq73CMMcZrPLp05Z4BrCyx4pEaAZsLvE8GuhRXR1VzRCQdiHLLfy20bSN3ubg2bwVmqOrWY90tJSIjgBEATZo0OY6P4x1v/vEmh3IOcXun28t938YY40ulm9C8ghGRhjiX4l4pqa6qjlXVRFVNjI4u3wf+N+/dzNTVU+kX349mEc3Kdd/GGONrvkwoKThD3ueLccuKrCMiQUAEzlP4xW1bXHlHoAWQJCIbgXARSfLWB/GWt5e/TWBAoHXEG2NOSL5MKAuAeBGJE5EQnE72wnOozMCZUhicJ/C/V2eClhnAIPcusDggHphfXJuq+oWqnqSqsaoaC2SoagsffrbjtiNjB5+t+4zLWlxGdLgNhWaMOfF41IdSGm6fyK3ALCAQGK+qK0RkNLDQnaBrHPCeezaxGydB4NabitNvkwPcoqq5AEW16avP4E3/W/k/cjWXoW2GllzZGGMqoRJnbDyRldeMjXuz9nLBtAvo1qgbz5z9jM/3Z4wxvlTcjI0nRKd8RTd97XQOZB/g2rY2pqYx5sRlCcXHVJVpa6bRPro9raNa+zscY4zxGUsoPrZo+yI27t3IFS2v8HcoxhjjU5ZQfOyjtR9RI7gGFzS9wN+hGGOMT1lC8aH0zHS+2fQNFze7mPDgcH+HY4wxPmUJxYe+3PAlmbmZ9I/v7+9QjDHG5yyh+NDMDTNpUbsFraJa+TsUY4zxOUsoPrL9wHYW71jMBbHWd2KMqRosofjIt399i6JcGHthyZWNMeYEYAnFR2ZtnEXLyJY2qrAxpsqwhOID2w5sY/GOxXZ2YoypUiyh+MC3m74FsGdPjDFViiUUH/gx+UfiIuKIjYj1dyjGGFNuLKF4WUZ2Bgu3L+SsRmf5OxRjjClXllC8bMG2BWTnZVtCMcZUOZZQvGxuylyqBVUjsf5RUwUYY8wJzRKKF6kqP6X8ROeTOhMSGOLvcIwxplxZQvGiTXs3kbI/xS53GWOqJJ8mFBHpJSKrRSRJREYVsT5URKa4638TkdgC6+53y1eLyIUltSkik9zy5SIyXkSCffnZivLzlp8BLKEYY6oknyUUEQkEXgN6A62Bq0Sk8JSFw4E0VW0BvAA85W7bGhgEtAF6Aa+LSGAJbU4CTgESgGrA9b76bMVZsG0BMTViiKkZU967NsYYv/PlGUpnIElV16tqFjAZ6FuoTl9ggrs8DeghIuKWT1bVTFXdACS57RXbpqp+qS5gPlCu3+qqyqLti+hUv1N57tYYYyoMXyaURsDmAu+T3bIi66hqDpAORB1j2xLbdC91XQPMLPMnOA7r09ezJ3OPJRRjTJV1InbKvw78qKpzi1opIiNEZKGILExNTfXaThdtXwRgtwsbY6osXyaUFKBxgfcxblmRdUQkCIgAdh1j22O2KSIPA9HAXcUFpapjVTVRVROjo6OP8yMVb+H2hdSrVs/6T4wxVZYvE8oCIF5E4kQkBKeTfUahOjOAoe7yFcD3bh/IDGCQexdYHBCP0y9SbJsicj1wIXCVqub58HMdpWD/idMFZIwxVU+QrxpW1RwRuRWYBQQC41V1hYiMBhaq6gxgHPCeiCQBu3ESBG69qcBKIAe4RVVzAYpq093lG8AmYJ77pf6xqo721ecrKGV/Cjsydlj/iTGmSvNZQgHnzivgy0JlDxVYPgQMKGbbx4HHPWnTLffpZzmW/P4TSyjGmKrsROyUL3dLUpdQM6QmzWrb7IzGmKrLEooXrNi5grZRbQkQO5zGmKrLvgHLKDM3k7V71tKmbht/h2KMMX5lCaWM1uxeQ05eDm2iLKEYY6o2SyhltGKXc5OZJRRjTFVnCaWMVuxaQZ2wOpxU/SR/h2KMMX5lCaWMlu9cTpuoNvZAozGmyrOEUgYZ2RmsT19vHfLGGIMllDJZnbaaPM2jbVRbf4dijDF+ZwmlDJbvXA5A66jC84YZY0zVYwmlDFbtXkV0tWiiw703arExxlRWllDKIGlPEvGR8f4OwxhjKgRLKKWUm5fL+j3raVG7hb9DMcaYCsESSikl70/mUO4hSyjGGOOyhFJKSWlJAHbJyxhjXJZQSmntnrUANIuwIeuNMQYsoZRa0p4kYmrEEB4c7u9QjDGmQrCEUkpJaUm0iLT+E2OMyWcJpRSycrPYuHcj8bWt/8QYY/L5NKGISC8RWS0iSSIyqoj1oSIyxV3/m4jEFlh3v1u+WkQuLKlNEYlz20hy2wzx1efakL6BXM21DnljjCnAZwlFRAKB14DeQGvgKhEpPEbJcCBNVVsALwBPudu2BgYBbYBewOsiElhCm08BL7htpblt+0TSHucOL7tl2Bhj/ubLM5TOQJKqrlfVLGAy0LdQnb7ABHd5GtBDnHHg+wKTVTVTVTcASW57RbbpbnOe2wZum5f56oMl7UkiSIKIrRXrq10YY0yl48uE0gjYXOB9sltWZB1VzQHSgahjbFtceRSwx22juH0BICIjRGShiCxMTU0txceCxjUb06dFH4IDg0u1vTHGnIiC/B1AeVPVscBYgMTERC1NG/3i+9Evvp9X4zLGmMrOl2coKUDjAu9j3LIi64hIEBAB7DrGtsWV7wJqu20Uty9jjDE+5MuEsgCId+++CsHpZJ9RqM4MYKi7fAXwvaqqWz7IvQssDogH5hfXprvNbLcN3DY/9eFnM8YYU4jPLnmpao6I3ArMAgKB8aq6QkRGAwtVdQYwDnhPRJKA3TgJArfeVGAlkAPcoqq5AEW16e7yPmCyiIwBFrttG2OMKSfi/Oe+akpMTNSFCxf6OwxjjKlURGSRqiYWLrcn5Y0xxniFJRRjjDFeYQnFGGOMV1hCMcYY4xVVulNeRFKBTaXcvC6w04vheIvFdXwsruNjcR2fihoXlC22pqoaXbiwSieUshCRhUXd5eBvFtfxsbiOj8V1fCpqXOCb2OySlzHGGK+whGKMMcYrLKGU3lh/B1AMi+v4WFzHx+I6PhU1LvBBbNaHYowxxivsDMUYY4xXWEIxxhjjFZZQSkFEeonIahFJEpFRfoyjsYjMFpGVIrJCRG53yx8RkRQRWeK+LvJDbBtFZJm7/4VuWR0R+UZE1rr/RpZzTCcXOCZLRGSviNzhj+MlIuNFZIeILC9QVuTxEcfL7u/bUhE5tZzjekZEVrn7ni4itd3yWBE5WOC4vVHOcRX7cxOR+93jtVpELiznuKYUiGmjiCxxy8vzeBX33eDb3zFVtddxvHCGzV8HNANCgD+A1n6KpQFwqrtcE1gDtAYeAe7283HaCNQtVPY0MMpdHgU85eef4zagqT+OF9AdOBVYXtLxAS4CvgIE6Ar8Vs5xXQAEuctPFYgrtmA9PxyvIn9u7t/AH0AoEOf+vQaWV1yF1j8HPOSH41Xcd4NPf8fsDOX4dQaSVHW9qmYBk4G+/ghEVbeq6u/u8j7gT6CRP2LxUF9ggrs8AbjMf6HQA1inqqUdKaFMVPVHnDmACiru+PQFJqrjV5zZSRuUV1yq+rWq5rhvf8WZEbVcFXO8itMXmKyqmaq6AUjC+bst17hERIArgQ98se9jOcZ3g09/xyyhHL9GwOYC75OpAF/iIhILdAR+c4tudU9dx5f3pSWXAl+LyCIRGeGW1VfVre7yNqC+H+LKN4gj/9D9fbyg+ONTkX7nrsP5n2y+OBFZLCJzRKSbH+Ip6udWUY5XN2C7qq4tUFbux6vQd4NPf8csoZwARKQG8BFwh6ruBf4LNAc6AFtxTrvL21mqeirQG7hFRLoXXKnOebZf7lkXZ/roPsCHblFFOF5H8OfxKY6IPIAzg+okt2gr0ERVOwJ3Ae+LSK1yDKnC/dwKuYoj/9NS7seriO+Gw3zxO2YJ5filAI0LvI9xy/xCRIJxfmEmqerHAKq6XVVzVTUPeAsfne4fi6qmuP/uAKa7MWzPP412/91R3nG5egO/q+p2N0a/Hy9XccfH779zIjIMuAS42v0iwr2ktMtdXoTTV9GyvGI6xs+tIhyvIKAfMCW/rLyPV1HfDfj4d8wSyvFbAMSLSJz7P91BwAx/BOJeox0H/KmqzxcoL3jt83JgeeFtfRxXdRGpmb+M06m7HOc4DXWrDQU+Lc+4Cjjif47+Pl4FFHd8ZgBD3DtxugLpBS5b+JyI9ALuBfqoakaB8mgRCXSXmwHxwPpyjKu4n9sMYJCIhIpInBvX/PKKy3U+sEpVk/MLyvN4FffdgK9/x8rjjoMT7YVzR8QanP9hPODHOM7COWVdCixxXxcB7wHL3PIZQINyjqsZzl02fwAr8o8REAV8B6wFvgXq+OGYVQd2AREFysr9eOEktK1ANs716uHFHR+cO29ec3/flgGJ5RxXEs719fzfsTfcuv3dn+8S4Hfg0nKOq9ifG/CAe7xWA73LMy63/F3gn4XqlufxKu67wae/Yzb0ijHGGK+wS17GGGO8whKKMcYYr7CEYowxxissoRhjjPEKSyjGGGO8whKKMWUgIrVF5OYC7xuKyDR/xmSMv9htw8aUgTtO0ueq2tbfsRjjb3aGYkzZPAk0d+e3eMad82I5OMOViMgn7rwTG0XkVhG5yx0c8FcRqePWay4iM92BNOeKyCmFdyIinUVknrvtLyJyslseLiJT3XkvpovIbyKS6K67Spw5aZaLyFPleExMFRXk7wCMqeRGAW1VtQMcPmMpqC3OSK9hOE+c36eqHUXkBWAI8CIwFuep6rUi0gV4HTivUDurgG6qmiMi5wNP4Dx5fTOQpqqtRaQtzhPRiEhDnLlLOgFpOCM/X6aqn3jvoxtzJEsoxvjWbHXmo9gnIunAZ275MqCdOxrsGcCHzvBLgDMxVGERwAQRiccZUiPYLT8LeAlAVZeLyFK3/DTgB1VNBRCRSTiTQX3ixc9mzBEsoRjjW5kFlvMKvM/D+fsLAPbkn+Ecw2M4yely9yzoB++GaUzZWR+KMWWzD2eK1VJRZ46KDSIyAA7P7d2+iKoR/D2c+LAC5T/jzAqIiLQGEtzy+cDZIlLXHeH2KmBOaeM0xhOWUIwpA3Xmt/jZ7fh+ppTNXA0MF5H80ZmLmlL6aeA/IrKYI68svA5Ei8hKYIy7ff7Q46OA2TijPi9SVX9NF2CqCLtt2JhKzD37CFbVQyLSHGdI8pNVNcvPoZkqyPpQjKncwoHZ7ux8AtxsycT4i52hGGOM8QrrQzHGGOMVllCMMcZ4hSUUY4wxXmEJxRhjjFdYQjHGGOMV/w9F02g7u+6pgAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "T = np.linspace(0, 200, 301)\n",
    "RAA, _ = my_history.coalescence_rate_trajectory(T, {\"R\": 2})\n",
    "RBB, _ = my_history.coalescence_rate_trajectory(T, {\"B\": 2})\n",
    "RAB, _ = my_history.coalescence_rate_trajectory(T, {\"R\": 1, \"B\": 1})\n",
    "\n",
    "plt.plot(T, RAA, label=\"two R samples\")\n",
    "plt.plot(T, RBB, label=\"two B samples\")\n",
    "plt.plot(T, RAB, label=\"one of each\")\n",
    "plt.legend()\n",
    "plt.xlabel(\"time ago\")\n",
    "plt.ylabel(\"coalescence rate\");"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See [the documentation](https://tskit.dev/msprime/docs/stable/demography.html#sec-demography-numerical) for more examples."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Other ancestry models\n",
    "\n",
    "(11.45am)\n",
    "\n",
    "When might `msprime`'s default (Hudson coalescent) model of genealogy not be ideal for you? Consider the following assumptions it makes:\n",
    "\n",
    " - Selective neutrality\n",
    " - Random mating and survival outcomes\n",
    " - Sample is small relative to population size\n",
    " \n",
    "However, it is possible to simulate from a selection of other models in `msprime` as well! \n",
    "The full list is [here](https://tskit.dev/msprime/docs/stable/ancestry.html#sec-ancestry-quickref)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Discrete-time Wright-Fisher\n",
    "\n",
    "DTWF simulations are less efficient than coalescent simulations, but should perform better\n",
    "\n",
    " - on analyses of very recent history (especially for haplotype-based information like identity-by-descent and local ancestry)\n",
    " - on small populations, or when sample size is a large fraction of the population size\n",
    " - when modelling very long range correlations across chromosomes is important\n",
    " \n",
    "See [this paper](https://doi.org/10.1371/journal.pgen.1008619) for a more comprehensive comparison between DTWF and coalescent simulations."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Consider the following (coalescent) simulation of two individuals in a population of two."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "                samples=2,\n",
    "                population_size=2,\n",
    "                random_seed=82,\n",
    "                sequence_length=20,\n",
    "                recombination_rate=0.3)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that we now have over 4 distinct chromosomes within a single generation, and indeed, some of these chromosomes inherit from multiple ancestors within a single generation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts.tables.nodes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is no longer a problem when we use a discrete-time Wright-Fisher simulation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DTWF sim\n",
    "ts = msprime.sim_ancestry(\n",
    "                samples=2,\n",
    "                population_size=2,\n",
    "                model=msprime.DiscreteTimeWrightFisher(),\n",
    "                random_seed=82,\n",
    "                sequence_length=10,\n",
    "                recombination_rate=0.2)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts.tables.nodes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simple selective sweeps\n",
    "\n",
    "We can also run simple simulations of genic selection using a structured coalescent model:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sweep_model = msprime.SweepGenicSelection(\n",
    "    position=1,  # start of chrom\n",
    "    start_frequency=0.2,\n",
    "    end_frequency=0.8,\n",
    "    s=0.25,\n",
    "    dt=1e-5,\n",
    ")\n",
    "\n",
    "ts = msprime.sim_ancestry(\n",
    "    samples=3,\n",
    "    sequence_length=5e7,\n",
    "    population_size=100,\n",
    "    model=sweep_model,\n",
    "    recombination_rate=1e-8,\n",
    "    random_seed=123456)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'd recommend using `SLiM` if selection is an important part of your demographic model."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mixing different ancestry models\n",
    "\n",
    "(11.52am)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can also combine multiple ancestry models inside a single `msprime` simulation. Here's a simple example involving a hybrid DTWF and coalescent simulation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "models=[msprime.DiscreteTimeWrightFisher(duration=500),\n",
    "        msprime.StandardCoalescent()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    10,\n",
    "    population_size=100,\n",
    "    sequence_length=1e6,\n",
    "    recombination_rate=1e-8,\n",
    "    model=models,\n",
    "    random_seed=6789\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
